---
title: 'Concatenate Gene Expression Files'
author: "Jenny Smith"
date: "Feb 14, 2022"
output: html_document
---


#Set-up 

```{r setup}
library(knitr)

# dir.create(file.path(BEATAML,"RNA/mRNAseq/analysis/2022.02.14_Cat_Kallisto_Quant"))
knitr::opts_knit$set(root.dir = file.path(BEATAML,"RNA/mRNAseq/analysis/2022.02.14_Cat_Kallisto_Quant"))
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, 
                      fig.align='center', 
                      fig.width = 10, fig.height = 10)

options(stringsAsFactors = FALSE)
table = function (..., useNA = 'ifany') base::table(..., useNA = useNA)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)

library(DeGSEA)
```


# Read in the Clinical Data 

```{r}
manifest <- read.csv(file.path(BEATAML,"RNA/mRNAseq/metadata/BEAT_AML_STAR-aligner_GCD_Data_Manifest_with_CDE.csv"))


head(manifest)
dim(manifest)
```


# Read in a Gene to Transcript ID Map 

https://www.gencodegenes.org/human/stats.html

```{r}
IDmap <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_Gene_IDmap.csv")

head(IDmap[,1:5])
dim(IDmap)
```

# TXimport with the HDF5 files

https://wurmlab.github.io/genomicscourse/2016-SIB/practicals/rnaseq/TP2
For this purpose, they introduced the "scaledTPM" values, which are obtained by summing the transcript-level TPMs by gene, and multiplying them with the total library size in millions. 

ScaledTPM values are artificial values, transforming underlying abundance measures to the scale of read counts. This allows to incorporate the information provided by the sequencing depth, and work with RNA-seq differential expression tools that were developed to use read counts.

countsFromAbundance:
character, either "no" (default), "scaledTPM", or "lengthScaledTPM", for whether to generate estimated counts using abundance estimates scaled up to library size (scaledTPM) or additionally scaled using the average transcript length over samples and the library size (lengthScaledTPM). if using scaledTPM or lengthScaledTPM, then the counts are no longer correlated with average transcript length, and so the length offset matrix should not be used.

```{r}
files <- data.frame(dir(file.path(BEATAML,"RNA/mRNAseq/level3/transcript/2020Jun_GDC_Kallisto_GRCh38_Illumina_data/"), 
             recursive = T, pattern = ".h5"))

head(files)
```

```{r}
library(rslurm)
```

```{r}
files <-patient.IDs$filename %>% 
  set_names(patient.IDs$Final_Colname)

tx2gene <- dplyr::select(IDmap, transcript_id, gene_id)

sopt <- list(nodes='1', 'cpus-per-task'='16',
             'partition'='largenode', 'mem'='62G',
             'time' = '24:00:00', 'mail-type'='END,FAIL',
             'mail-user'='jlsmith3@fredhutch.org') 

```

```{r}
txi.geneLevel.job <- slurm_call(f=tximport,
                     jobname = "DSAML_gene",
                     params =  list(files = files,
                                    type="kallisto", 
                                    tx2gene = tx2gene, 
                                    txIn = TRUE,
                                    txOut = FALSE,
                                    ignoreAfterBar = TRUE, 
                                    dropInfReps= TRUE,
                                    countsFromAbundance = "scaledTPM"),
                     add_objects = c("files","tx2gene"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 


```

```{r}
get_job_status(txi.geneLevel.job)
```

```{r}
txi.txLevel.job <- slurm_call(f=tximport,
                     jobname = "DSAML_Tx",
                     params =  list(files = files,
                                    type="kallisto", 
                                    tx2gene = tx2gene, 
                                    txIn = TRUE,
                                    txOut = TRUE,
                                    ignoreAfterBar = TRUE, 
                                    dropInfReps= TRUE,
                                    countsFromAbundance = "scaledTPM"),
                     add_objects = c("files","tx2gene"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 


# str(txi.txLevel.job) 
```

```{r}
get_job_status(txi.txLevel.job)
```


## read in the results of tximport 

```{r}
txi.geneLevel <- readRDS("_rslurm_DSAML_gene/results_0.RDS") 
txi.geneLevel$countsFromAbundance
```

```{r}
lapply(txi.geneLevel[c(1:3)], function(x) head(x[,1:5]))

sapply(txi.geneLevel[c(1:3)], dim)
```

```{r}
# saveRDS(txi.geneLevel$abundance,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_Abundance_TPM.RDS")
# 
# saveRDS(txi.geneLevel$counts,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_scaledTPM_counts.RDS")
# 
# saveRDS(txi.geneLevel$length,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_scaledTPM_length.RDS")
```


```{r}
txi.transcriptLevel <- readRDS("_rslurm_DSAML_Tx/results_0.RDS") 
txi.transcriptLevel$countsFromAbundance
```

```{r}
new_rownames <- str_split_fixed(rownames(txi.transcriptLevel$abundance), 
                                pattern = "\\|", n = 2)[,1] 

txi.transcriptLevel[c(1:3)] <- lapply(txi.transcriptLevel[c(1:3)],set_rownames, value=new_rownames)
```

```{r}
lapply(txi.transcriptLevel[c(1:3)], function(x) head(x[,1:5]))
lapply(txi.transcriptLevel[c(1:3)], function(x) tail(x[,1:5]))

sapply(txi.transcriptLevel[c(1:3)], dim)
```

```{r}
# saveRDS(txi.transcriptLevel$abundance,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_Abundance_TPM.RDS")
# 
# saveRDS(txi.transcriptLevel$counts,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_scaledTPM_counts.RDS")
# 
# saveRDS(txi.transcriptLevel$length,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_scaledTPM_length.RDS")
```



#Session Information 

```{r}
sessionInfo()
```

