---
title: 'GDC data download'
author: "Jenny Smith"
date: "February 2, 2018"
output: html_document
---


#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE,
                      fig.align='center', fig.height=5, fig.width=8, dpi = 600)
knitr::opts_knit$set(root.dir = file.path(SCRATCH,"jlsmith3"))
options(stringsAsFactors = FALSE)
```


```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
getwd()
```


#References

https://bioconductor.org/packages/release/bioc/vignettes/GenomicDataCommons/inst/doc/overview.html#quickstart
http://bioconductor.org/packages/release/bioc/vignettes/TCGAutils/inst/doc/TCGAutils.html

http://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/index.html
http://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/query.html#get_manifest_file


#Installation  

```{r}
# BiocManager::install("TCGAbiolinks") #as back up 
# library(TCGAbiolinks)
```

```{r message=FALSE}
# BiocManager::install('GenomicDataCommons')
library(GenomicDataCommons)
```

```{r}
GenomicDataCommons::status()
```

```{r}
stopifnot(GenomicDataCommons::status()$status=="OK")
```


# Orientation with Package 

There are 4 main types of objects:
1. projects()
2. files()
3. cases() 
4. annotations()

Each of these can be filtered and specific fields can be selected. 
Then piped (`%>%`) into `results()` or `results_all()` to get a list with each file/case ID and its metadata (like date sumbitted to GDC)


```{r}
available_fields('projects') %>% length()
available_fields('files') %>% length()
available_fields('cases') %>% length()
available_fields('annotations') %>% length()
```

## Projects 

```{r}
pQuery = projects()
head(projects())
# length(pQuery) #5 
```

```{r}
default_fields(pQuery)
```

```{r}
presults = pQuery %>% results() #results_all()
class(presults)
str(presults)
ids(presults) #10 datasets returned by default when no arguments given to projects and no filters added
```


## Files 

```{r}
default_fields(files()) #data types - eg RNA-seq, DNAme, miRNA-seq, etc. 

files() %>% count() #without filters, and only 10 Cohorts has 358,679 files
```


## Cases 

```{r}
qcases = cases() #patient samples
qcases$fields #grep_fields(), field_picker() can be useful here
head(available_fields('cases')) #clinical data elements "columns"

qcases %>% count() #33,096 patient cases/samples
```



# Summarize Available Information 


```{r}
# total number of files of a specific type. 
#remember to use available_feilds(files()) for your facet arguments. 
res = files() %>% 
  facet(c('type','data_type','data_format',
          'cases.project.project_id')) %>% #Default is to set facets for all default fields. 
  aggregations()

# res$type
# res$data_type
# res$data_format
res$cases.project.project_id$key %>%
  grep("TARGET", ., value=TRUE)
```

```{r}
res.projects <- projects() %>%
  facet(c("project_id")) %>%
  aggregations() 

# str(res.projects)
```

```{r}
res.analysis <- files() %>%
  facet(grep_fields('files','analysis.workflow')) %>% 
  aggregations()

# res.analysis$analysis.workflow_version
# res.analysis$analysis.workflow_type
```

```{r}
res.cases <- cases() %>% 
  facet() %>% #Default is to set facets for all default fields. 
  aggregations()

head(res.cases$primary_site)
```


# Query GDC for Gene Expression Data and Clinical Data

```{r}
# grep('project',available_fields('files'),value=TRUE) #see facets above
# grep('type',available_fields('files'),value=TRUE) 
# grep('cases.samples.submitter_id',available_fields('files'), value=TRUE)
```

 
```{r}
#Note: Here I need the sample UUID because it will map file ID to TARGET Barcode later, and is not included in the default fields. 
#so I used some selected fields, such as project ID, associated_entities (Sample UUIDs),and workflow type, in addition to  the default.

qfiles <- files(fields=c("cases.project.project_id",default_fields('files'), grep_fields('files', "associated_entities"), 
                         "analysis.analysis_type", "analysis.workflow_type", "analysis.workflow_version")) %>% 
  filter(~ type == 'gene_expression' & 
           analysis.workflow_type == 'HTSeq - Counts' &
          (cases.project.project_id == "TARGET-AML" |
             cases.project.project_id == "TARGET-NBL" | 
             cases.project.project_id == "TARGET-WT")) 


qfiles %>% count() #473
# head(qfiles)
```

```{r}
qfiles_AML <- files(fields=available_fields('files')) %>% 
  filter(~ type == 'gene_expression' & 
           # grepl("TARGET", cases.project.project_id ) & #doesn't work
           cases.project.project_id == "TARGET-AML" &
           analysis.workflow_type == 'HTSeq - Counts')


qfiles_AML %>% count() #187 
str(qfiles)
# qfiles$fields
```


#Create ID Map for Manifest, Clinical, and Expression Files

```{r}
res.expn <-  qfiles %>%
  results_all()

length(res.expn)
names(res.expn)
# str(res.expn)
# lapply(res.expn,head)
```

The "cases" feilds - if used in `results()` contains much of the clinical data elements that are extracted through `gdc_clinical()` function used below. 

```{r}
#there are 2 Sample IDs associated with a single file. 
#Filter these out, since they are suspect. 
idx <- sapply(res.expn$associated_entities , nrow) %>% grep(2, .)
```

```{r}
#check that these are the expected files/samples which more than 1 ID associated

# res.expn$associated_entities[idx]
# res.expn$cases[idx]
```

```{r}
ID.map <- res.expn[ !grepl("list|data.frame", sapply(res.expn, class)) ] %>%
  as.data.frame() %>%
  slice(-idx) %>% #remove the multiple ID mapped files
  mutate(project.project_id=unlist(res.expn$cases[-idx])) %>%
  bind_cols(., bind_rows(res.expn$associated_entities[-idx])) #associated_entities was a list of data.frames, converted a single data.frame with bind_rows()

head(ID.map)
dim(ID.map) # 469  20
```

```{r}
table(ID.map$project.project_id)
```


#Create a Download Manifest file

```{r}
qfiles <- qfiles %>%
  filter(~ file_id %in% names(res.expn$associated_entities[-idx]) )

qfiles %>% count() #469
```


```{r}
manifest_df = qfiles %>% manifest()
head(manifest_df)
dim(manifest_df) #469 
```


#Download the files 

```{r}
#Add the GDC Client to your R options
options(gdc_client="/home/jlsmith3/scripts/opt/bin/gdc-client")
gdc_client()
```

```{r}
#Set your destination directory
gdc_set_cache(directory = "/fh/scratch/delete90/meshinchi_s/Hackathon_2019")
```

```{r}
#download step
fnames = gdcdata(manifest_df$id[1:2],progress=FALSE,access_method = "api", use_cached = FALSE)

head(fnames)
```


#Reorganize the Downloaded Files and Check the MD5 Sums

Using a little bash here. 

```{r}
system("ls -d")
system("pwd")
```

```{r}
system("find $PWD -type f -name '*.htseq.*' -exec mv {} . \;")
```

```{r}
#check MD5 sums
```



#Download Clinical Data 

https://seandavi.github.io/2018/03/extracting-clinical-information-using-the-genomicdatacommons-package/

##from GDC Database

```{r}
cases() %>% filter(~ project.project_id == "TARGET-AML") %>% count() #988 patients

case_ids <- cases() %>%
  filter(~ project.project_id == "TARGET-AML") %>%
  ids()

head(case_ids)
```

```{r}
#download the clinical data per patient
clin_res = gdc_clinical(case_ids)

sapply(clin_res, dim) %>% 
  set_rownames(c("rows","colunms"))

# sapply(clin_res, colnames)
# sapply(clin_res, head)
```

```{r}
#Combine all Clinical Data into a Master File
full_clin = with(clin_res,
     main %>% #what is main here? not a function...
     left_join(demographic, by = "case_id") %>%
     left_join(exposures, by = "case_id") %>%
     left_join(diagnoses, by = "case_id"))

head(full_clin)
dim(full_clin) # 988  33
```

## Precompiled Clinical Data Elements from Study Authors 

```{r}
library(GenomicDataCommons)
```

```{r}
qClin <- files() %>%
    filter(~ type == 'clinical_supplement' &
             cases.project.project_id == "TARGET-AML") 

# str(qClin)
# qClin %>% count()
```

```{r message=FALSE}
manifest.clin <- manifest(qClin)
head(manifest.clin)

write.csv(manifest.clin, "TARGET_AML_Clinical_Data_on_GDC_Manifest.csv", row.names = FALSE)
```

```{r}
#Add the GDC Client to your R options
options(gdc_client=file.path(dirname(HOME), "scripts/opt/bin/gdc-client"))
gdc_client()
```

```{r}
#Set your destination directory
gdc_set_cache(directory = file.path(SCRATCH, "jlsmith3"))
gdc_cache()
```

```{r}
#Download the data sets
clinFile <- gdcdata(uuids = manifest.clin$id)
clinFile
```







#Session Information 

```{r}
sessionInfo()
```

