---
  title: "Concatenate Intron Level 1031 RNA-seq"
author: "Jenny Smith"
date: "August 13, 2018 "
output: html_document
---
  
  
```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height = 5, fig.width = 5)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2017.10.09_Concatenate_1031_RNAseq/')
```

```{r message=FALSE}
library(dplyr)
library(ggplot2)
library(stringr)
library(reshape2)
getwd()
```

```{r}
source("~/scripts/conversion_scripts/Merge_Cat_FixDupIDs_Function.r")
source("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/rmDupGenes_Function.r")
```




#Define Functions to be used. 

```{r}
#Function for the TPM conversion. 
# Based on https://groups.google.com/forum/#!topic/rsem-users/W9RQrZIOzA4
#Useage: sapply(cated$RPKM, RPKM_to_TPM)

RPKM_to_TPM <- function(RPKM){
  conversionFactor <- sum(RPKM) / 1E6
  TPM <- RPKM / conversionFactor
  return(TPM)
}
```


```{r}
addCols <- function(df,symbol,Intron){
  library(dplyr)
  
  df <- as.data.frame(df)
  
  df <- df %>%
    mutate(geneSymbol=symbol,
           intron=Intron) %>%
    select(geneSymbol, intron, everything())
  
  return(df)
}
```



#Identify Files to Be Concatenated

```{r}
filepath <-  "/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/level3/intron/2017July_BCCA_1031_Illumina_data/"

allfiles <- paste0(filepath, dir(path = filepath, 
                                 pattern = ".normalized"))

allfiles <- allfiles[-grep("Kas|MV4", allfiles)] #remove cell lines
head(allfiles)
```


```{r}
length(allfiles) #1111 AML files 
```



#Columns Description

The .intron.ratios.normalized file contains:
-*column 1*: gene id
-column 2: transcript id or "merged_<gene_id>" if the analysis was performed in
collapse mode
-colomn 3: strand of the annotation ("+" or "-")
-*column 4*: intron number (relative to the gene's strand)
-column 5: "intron" (annotation type)
-column 6: chromosome
-column 7: start of the intron (relative to the positive strand, so start<end)
-column 8: end of the intron (relative to the positive strand, so start<end)
-*column 9*: number of fractional reads inside this intron's boundaries (sum of the
fraction of each read inside this region)
-column 10: total coverage across this intron (sum of the coverage depth at each
base in this intron)
-column 11: collapsed gene or transcript length
-column 12: average coverage of the collapsed gene or transcript
-column 13: intron length
-column 14: average coverage inside this intron-- total coverage across this
intron divided by the intron's length
-column 15: intron-to-gene coverage ratio: average intron coverage divided by the
average collapsed gene or transcript coverage
-column 16: duplicate of column 14
-*column 17*: normalized average coverage inside this intron (RPKM) -- (number of
this intron's fractional reads x 1000000000) / (NORM_TOTAL x intron length)



#Begin Concatenation 

```{r}
#Pattern to select the Target Barcode
pattern <- "^.+\\/([BPR][A-Z0-9].+R)\\_.+"

```

```{r}
#Select the column indices 
selected <- c(1,4,9,17)
```

```{r results='hide'}
cated <- catExpnData(filenames = allfiles,regex =  pattern, cols = selected, header = FALSE)
names(cated) <- c("geneSymbol","Intron_Number", "Fractional_Reads", "RPKM")
```

```{r}
lapply(cated, function(x) head(x[,1:5]))

lapply(cated,dim)
```



#Check that the Gene Ids are in the Same Order


```{r}
all(apply(cated$geneSymbol,2, function(x) identical(x=x,y=cated$geneSymbol[,1])))
```

```{r}
all(apply(cated$Intron_Number,2, function(x) identical(x=x,y=cated$Intron_Number[,1])))
```


#Add TPM Values 

```{r}
cated[["TPM"]] <- apply(cated$RPKM,2, RPKM_to_TPM)
```

```{r}
# apply(cated[["TPM"]],2,sum) #all 1 million
```


#Add columns for the Gene Symbol and Ensembl Symbol 

```{r}
withCols <- lapply(cated[3:5], addCols, 
                   symbol = cated$geneSymbol[,1], Intron = cated$Intron_Number[,1])


lapply(withCols, function(x) head(x[,1:5]))
```



#Save the output

```{r}
# write.csv(withCols, "TARGET_AML_1031_DS.AML_HTSeq_Counts.csv", row.names = FALSE)
```

```{r}
# lapply(names(withCols), function(x) write.csv(withCols[[x]],paste0("TARGET_AML_AAML1031_RNASeq_Intron_",x,".csv"), row.names = FALSE))
```


NOTE:
  
  If markdown cannot be completed with pandoc (memory issues), use the following

> library("rmarkdown")
> render("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/TARGET_AML_1031_Cat_RNAseq.Rmd", "html_document")


#Session Info

```{r}
sessionInfo()
```

