---
title: "TARGET_AML_AAML0531_TallyRNAseq.Rmd"
author: "Jenny Smith"
date: "September 1, 2017"
output: html_document
---


```{r}
#set working directory
# dir.create("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/RNAseq/analysis/2017.09.01_TallyRNAseq_0531")
```


```{r}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height = 4, fig.width = 6, message = FALSE)
knitr::opts_knit$set(root.dir = "/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2017.09.01_TallyRNAseq_0531/")
knitr::opts_chunk$set(cache=FALSE)
```

```{r}
library(magrittr)
library(stringr)
```


#Read in the clinical Data

```{r}
CDE <- read.csv("~/reference_mapping-files/TARGET_AML_current_asof_june30_2016_UPDATED_RhondaEdit_2017.06.08.csv", stringsAsFactors = FALSE, row.names = 1)
CDE <- subset(CDE, Protocol == "AAML0531")
head(CDE[,1:5]) #825 in AAML0531
```


```{r}
CDE.Entire0531 <- read.csv("~/reference_mapping-files/AAML0531_pts_with_relapse_info.csv", stringsAsFactors = FALSE)
head(CDE.Entire0531[,1:5])

# dim(CDE.Entire0531) #There are 1022 patient Reg numbers 
```



#Read in the current tally

```{r}
#Tallied RNAseq Samples with CDE (TARGET_AML_current_asof_June2016_FINAL.csv)
RNAseq.w.CDE <- read.csv("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/metadata/TARGET_AML_AAML0531_mRNAseq_DataAvailability.csv", stringsAsFactors = FALSE)
head(RNAseq.w.CDE[,1:3])
```


#Mapping File

```{r}
#USI to Reg for all 0531 patients from Rhonda, emailed on 9/1/17
ID.map <- read.csv("~/reference_mapping-files/USI_Reg_Num_conversion_0531.csv", stringsAsFactors = FALSE)
head(ID.map)
# dim(ID.map) #787 IDs
```


#Visualize the overlap 

```{r fig.heigh=8, fig.width=6}
#Create a venn diagram
library(VennDiagram)

v <- venn.diagram( list(CDE.0531=CDE$Reg., 
                        CDE.Entire0531=CDE.Entire0531$Reg.., 
                        ID.map=ID.map$reg_no),
                   filename = NULL,
                   units = "in",
                   height = 8,
                   width = 6,
                   cat.pos= 0,
                   cat.dist = -0.035,
                   col=c("lightgoldenrod", "deepskyblue", "blue"),
                   fill=c("lightgoldenrod", "deepskyblue", "blue"),
                   alpha=0.5,
                   main="TARGET AML AAML0531")

grid.draw(gList(v))
```


```{r results='hide'}
library(gplots)
table <- venn(list(CDE.0531=CDE$Reg., 
                        CDE.Entire0531=CDE.Entire0531$Reg.., 
                        ID.map=ID.map$reg_no), show.plot = FALSE)

thirtySeven <- attr(table, "intersections")
# thirtySeven$CDE.Entire0531
```


#Combining the ID maps

```{r}
ID.map <- merge(ID.map[,1:2], CDE.Entire0531[,1:2], by.x="reg_no", by.y="Reg..")
sel <- CDE[,c("Reg.", "TARGET.USI.1","Protocol")] %>% set_names(c("reg_no","USI", "Study"))
ID.map <- unique(rbind(ID.map,sel))
```

```{r}
#There are a total of 895 AAML0531 patient samples
dim(ID.map)
```

```{r}
# write.csv(ID.map, "TARGET_AML_0531_RegToUSI_IDmap.csv", row.names = FALSE)
```


#AAML053 Samples with Gene-Level RNAseq

```{r}
#select only patients who are known to have at least gene-level RNAseq expression. 
RNAseq.yes <- subset(RNAseq.w.CDE, RNAseq.w.CDE$geneLevelExpn == 1)
```


#AAML0531 Samples without Gene-Level RNAseq

```{r}
#setdiff to find those IDs in 0531 that do not have RNAseq at the gene-level.  
IDs.Noseq <- setdiff(ID.map$USI, RNAseq.yes$USI)

#474 Patients 
length(IDs.Noseq)
```

```{r}
IDs.Noseq[1:10]
```

```{r}
RNAseq.No <- subset(ID.map, ID.map$USI %in% IDs.Noseq)
# write.csv(RNAseq.No,"TARGET_AML_AAML0531_No_RNAseq_DataAvailable.csv", row.names = FALSE)
```


##AAML0531 with RNAseq

```{r}
#Intersect to find patients with RNAseq
IDs.withSeq <- intersect(RNAseq.yes$USI, ID.map$USI)

#511 have gene-level RNAseq data from AAM053
length(IDs.withSeq)
```


#Merge the RNAseq Tally with the CDE

```{r}
RNAseq.CDE <- merge(CDE, RNAseq.w.CDE, by.x="TARGET.USI.1", by.y="USI")
head(RNAseq.CDE[,1:5])
```


```{r}
library(magrittr)
RNAseq.CDE$Fusion.Cluster.Category <- RNAseq.CDE$Fusion.Cluster.Category %>% ifelse(. == "", "None", .)
unique(RNAseq.CDE$Fusion.Cluster.Category)
```


#Plot the Number of Variants with CDE

```{r}
library(pryr)

theme_clinData %<a-% { theme(plot.title = element_text(hjust = 0.5, size = 18),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_line(color = "gray"),
                       panel.grid.minor = element_blank(),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 0,hjust=0.5,vjust =0.5, size = 15),
                       axis.text.y = element_text(size = 16),
                       axis.title = element_text(size = 16),
                       legend.text = element_text(size=14),
                       legend.title = element_text(size=14))
}
```


```{r fig.width=12, message=FALSE}
library(ggplot2)
ggplot(RNAseq.CDE, aes(x=Fusion.Cluster.Category, fill=as.factor(geneLevelExpn))) +
  geom_bar(stat="count", position = position_dodge()) + 
  scale_y_continuous(limits = c(0,200), breaks = seq(0,200,by=10)) +
  scale_fill_manual(name="Gene-Level RNAseq", values = c("0"= "firebrick1", "1"="dodgerblue"), labels=c("0"="No", "1"="Yes")) +
  coord_flip() +
  theme_clinData + 
  labs(title="RNAseq Available for AAML0531")
```


```{r results='asis', fig.align="center"}
library(knitr) 
t <- table(RNAseq.CDE$Fusion.Cluster.Category, RNAseq.CDE$geneLevelExpn) 
dimnames(t)[[2]] <- c("No", "Yes") 
kable(t, format = "html", caption = "AAML0531 RNA-Seq Data Availablity.")  
```


