---
title: 'Concatenate TransAbyss Fusions'
author: "Jenny Smith"
date: 'September 23,2019'
output: html_document
---


#Set-up

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.width = 10, fig.height = 10)
knitr::opts_knit$set(root.dir = file.path(PROJHOME,'2019.06.07_Concatenate_Relapse_TransAbyss_Fusion'))
options(stringsAsFactors = FALSE)
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
library(purrr)
getwd()
```

```{r}
source("~/scripts/conversion_scripts/Merge_Cat_FixDupIDs_Function.r")
```


# Define Functions

```{r}
##### Define Functions ################

collapseRows <- function(col, uniq=FALSE){
  #designed for dplyr so that col is the vector of the column name
  if (uniq){col <- unique(col)}
  
  collapsed <- ifelse(all(is.na(col)), NA, paste(col, collapse = "; "))
  return(collapsed)
}



filterFusions <- function(Fusion, breakpoint,df2.Fusion,df2.Breakpoints){

  fus <- paste(Fusion,breakpoint,sep=" ")
  fus2 <- paste(df2.Fusion, df2.Breakpoints, sep=" ")
  
  if (fus %in% fus2){
      res <- 1 #these fusions will be filtered out. 
  }else{
    res <- 0 #these one will not. 
  }
  
  return(res)
}


removeDups <- function(col.rowwise){
  
  uniqued <- str_split(col.rowwise, "; ") %>% 
  unlist() %>%
  gsub(" ","",.) %>%
  unique() %>%
  paste(.,collapse = "; ")
  
  return(uniqued)
  
}

```




#Read in the CDEs 

```{r}
manifest <- read.csv("TARGET_AML_RBD_Relapse_Manifest_v3.csv") %>% 
  select(GSC.library=GSC.library,everything())

head(manifest)
dim(manifest)
```

```{r}
merged <- read.csv(file.path(CDE,"Merged/TARGET_AML_0531_1031_merged_CDEs_9.4.19.csv"))

merged <- merged %>% 
  filter(!is.na(USI))

dim(merged)

```


#Rename TransAbyss 

```{bash eval=FALSE} 
cd "/fh/scratch/delete90/meshinchi_s/2019.04_BCCA_mRNAseq_Relapse_Data_Downloads"

TA=$(find . -type f -name "*.tsv" )

for file in $(echo "$TA")
do 
  dir=$(dirname $file); 
  id=$(echo "$dir" | sed -E 's/^.+(B0[0-9]{4}).+/\1/' ); 
  orig=$(basename $file); echo $file; 
  cp -n $file $dir/${id}_$orig 
done
```



#Select the Files for Merging 

```{r}
path <- file.path(SCRATCH,"2019.04_BCCA_mRNAseq_Relapse_Data_Downloads")
```

```{r}
allfiles <- dir(path = path,
                recursive = TRUE,
             pattern = "B[0-9].+tsv", 
             full.names = TRUE) 

length(allfiles) #3801
head(allfiles,n=7)
```

```{r}
filetypes <- c("antisense_fusion","ITD","local","LSR","PTD","_sense_fusion","events_exons_novel")
forMerge <- lapply(filetypes, function(x) grep(x, allfiles, value = TRUE)) %>% 
  set_names(filetypes)

lapply(forMerge, length)
# lapply(forMerge, head)
```


# Begin Merging the Fusion Data Sets

```{r message=FALSE}
merged.datasets <- lapply(forMerge,catRbind,
                          regex="^.+(B[0-9]{5}).+",header=TRUE,ID.Col.Name="GSC.library")

#reorder the columns
merged.datasets <- lapply(merged.datasets,
                          function(x) select(x, GSC.library, everything())) 

# summary(merged.datasets)
sapply(merged.datasets, function(x) dim(x))
```


```{r}
add_USIs <- function(df,ref.df){
  df <- df %>% 
    left_join(., ref.df, by=c("GSC.library")) %>%
    select(Patient=Sample.ID,everything())
}

#This batch needs to add in the USIs 
merged.datasets <- lapply(merged.datasets, add_USIs,
                           ref.df=select(manifest,GSC.library,Sample.ID))
```


```{r}
#Updated the code for catRbind to include a dummy data.frame if there are no called fusions for that patient. 
names(merged.datasets) <- c("antisense_fusion","ITD","local","LSR","PTD","sense_fusion","events_exons_novel")

sapply(merged.datasets,function(x) length(unique(x$Patient)))
lapply(merged.datasets, function(x) head(x[,1:5]))
sapply(merged.datasets, function(x) dim(x))
```

```{r}
# lapply(names(merged.datasets), function(x)
#   write.csv(merged.datasets[[x]], file=paste0("TARGET_AML_0531_1031_TransAbyss_", gsub("^_", "",x), "_10.30.19.csv"), 
#             row.names = FALSE))
```


#Delete Temporary Renamed Files

```{bash eval=FALSE} 
cd "/fh/scratch/delete90/meshinchi_s/2019.08_BCCA_mRNAseq_DSAML_Data_Downloads/trans-abyss"
find . -type f -name "^B[0-9].+tsv" -exec rm {} +
```



# Read in the Sense Fusions Files

```{r}
sense <- read.csv("Originals/TARGET_AML_0531_1031_TransAbyss_sense_fusion_10.30.19.csv")

head(sense)
dim(sense) #23,339    36

length(unique(sense$Patient)) #543
```

```{r}

```


#Reformat and Clean Fusion Cols

```{r}
reformat.fusions <- function(df, manifest){
  # customCols <- c("Patient", "Fusion","All.Sense.Fusions.Found","All.Fusion.Exons","All.Genomic.Breakpoint.ReadPairs","All.Contig.Spanning.Reads", "All.Reading.Frames","All.BreakPoints", "All.Fusion.Sizes")
  # originalCols <- colnames(df)
  
  
  df <- df %>% 
    arrange(Patient) %>%
    
    mutate(Fusion.Detected.TransAbyss=ifelse(is.na(genes), "NoneDetected","Detected"), 
           USI=str_split_fixed(Patient,"-", n=5)[,3]) %>%
    mutate_at(vars(c("X5.gene","X3.gene")), funs(gsub("-","\\.", .))) %>% #replace dashes with a dot. 
    
    mutate(Fusion=ifelse(! is.na(genes), paste(X5.gene, X3.gene, sep="-"), NA),
           Exons=ifelse(! is.na(genes), paste(paste0(X5.gene,"exon",X5.exon), 
                                              paste0(X3.gene,"exon",X3.exon), sep="-"), NA),
           FusionFrame=ifelse(! is.na(genes), paste(Fusion, frame,"frame", sep="_"), NA),
           Type=ifelse(! is.na(genes), ifelse(grepl("^[0-9]", size), 
                                              "intrachromosomal", "interchromosomal"), NA)) %>%
    
    mutate(b1=str_split_fixed(breakpoint,"\\|", n=2)[,1], 
           b2=str_split_fixed(breakpoint,"\\|", n=2)[,2]) %>% 
    mutate(b1=paste(b1,str_split_fixed(genes,",", n=2)[,1], sep="_"), 
           b2=paste(b2,str_split_fixed(genes,",", n=2)[,2], sep="_")) %>% 
    
    #Needed to rearrange breakpoints to be in same orientation at X5.gene, X3. gene
    #Original TransAbyss data has breakpoints sorted by genomic coordinate, not which partenr was 5' or 3'
    rowwise() %>%
    mutate(X5.breakpoint = case_when(
              str_detect(b1, X5.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X5.gene) ~ str_split_fixed(b2, "_",n=2)[,1])) %>% 
    mutate(X3.breakpoint = case_when(
              str_detect(b1, X3.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X3.gene) ~ str_split_fixed(b2, "_",n=2)[,1])) %>% 
    
    rowwise() %>%
    select(Patient,USI,GSC.library,Fusion,Exons,
           X5.breakpoint,X3.breakpoint, FusionFrame,everything(),-b1,-b2) %>%
    mutate_at(vars(3:NCOL(.)), funs(as.character(.))) %>%
    mutate_at(vars(3:NCOL(.)), funs(ifelse(grepl("NoneDetected", Fusion.Detected.TransAbyss), 
                                           "NoneDetected", .))) %>% 
    ungroup()
   
  
 
  return(df)
}
```

```{r}
sense.fmt <- reformat.fusions(sense)
head(sense.fmt)
```

```{r}
length(unique(sense.fmt$Patient)) #543 patients
dim(sense.fmt) 
# write.csv(sense.fmt, "TARGET_AML_0531_1031_Relapse_TransAbyss_sense_fusions_reformatted_10.30.19.csv", row.names = FALSE)
```


#Filter out Fusions found in NBMs 

```{r}
NBM1 <- read.csv(file.path(PROJHOME,"2018.02.05_TransAbyss_Fusion/Batch1/Originals/TARGET_AML_1031_TransAbyss_sense_fusion_4.6.18.csv")
                   , row.names = 1) %>% 
  filter(grepl("^BM|^RO",Patient)) #Need NBMs only

# head(NBM1)
dim(NBM1) # 1520   35
```

```{r}
NBM2 <- read.csv(file.path(PROJHOME,"2018.02.05_TransAbyss_Fusion/Batch2/Originals/TARGET_AML_0531_TransAbyss_sense_fusion_10.02.18.csv"),
                   row.names = 1) %>% 
  filter(grepl("BM[0-9]|RO[0-9]",Patient)) #Need NBMs only

# head(NBM2)
dim(NBM2) #610  35
```

```{r}
NBM3 <- sense %>% 
  filter(grepl("BM[0-9]{4}|R0[0-9]",Patient)) %>% #Need NBMs only
  select(-GSC.library)

# head(NBM3)
dim(NBM3) #385  35
```

```{r}
#Using both NBM sets to account for the total fusions found in the NBM dataset
NBM <- NBM1 %>%
  bind_rows(.,NBM2, NBM3) %>%
  mutate_at(vars(X5.gene, X3.gene), ~gsub("-","\\.", .)) %>% #some readthrough genes included
  mutate(Fusion=ifelse(! is.na(genes), paste(X5.gene, X3.gene, sep="-"), NA)) %>%
  mutate_at(vars(4:NCOL(.)),funs(gsub("/", "-", .))) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(4:NCOL(.)),funs(gsub(" ", "", .))) %>% ##to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(4:NCOL(.)), funs(gsub("@","", .))) #to address the TRA-TRA@ , not sure how to find which gene was supposed to be TRA@??



dim(NBM) # 2515   37
length(unique(NBM$Patient)) #100
# write.csv(NBM,"~/TARGET_Fusions_in_NormalMarrrows_Trans-Abyss_4.27.20.csv",
#           row.names = FALSE)
```

```{r}
NBM.Fusions <- table(NBM$Fusion) %>%
  as.data.frame() %>%
  arrange(desc(Freq)) #%>%
  # filter(grepl("NPHP3\\.ACAD11-NPHP3", Var1))

head(NBM.Fusions)
```

```{r warning=FALSE}
sense.fmt.filter <- sense.fmt %>%
  mutate_at(vars(4:NCOL(.)),~gsub("/", "-", .)) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(4:NCOL(.)),~gsub(" ", "", .)) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(4:NCOL(.)), ~gsub("@","", .)) %>%
  
  rowwise() %>%
  mutate(FailedFilterFusions=filterFusions(Fusion, breakpoint, NBM$Fusion,NBM$breakpoint)) %>%
  ungroup() %>%
    
  group_by(Patient) %>%
  mutate(AllFail=ifelse(sum(FailedFilterFusions) == n(),"Yes","No")) %>%
  mutate(Fusion.Detected.TransAbyss=ifelse((AllFail == "Yes" & Fusion.Detected.TransAbyss != "NoneDetected"),
                                           "NonePassedFilter",Fusion.Detected.TransAbyss)) %>%
  ungroup()  %>%
  
  filter(AllFail == "Yes" | FailedFilterFusions == 0) %>%

  rowwise() %>%
  mutate_all(funs(as.character(.))) %>% #change all to characters so that numeric cols can be filled below with a charcter string.
  mutate_at(vars(Fusion:dgv,Fusion.Detected.TransAbyss:Type), funs(ifelse(Fusion.Detected.TransAbyss == "NonePassedFilter","NonePassedFilter",  .))) %>%
  mutate_at(vars(c("flanking_pairs","spanning_reads")), 
            funs(ifelse(Fusion.Detected.TransAbyss == "NonePassedFilter", NA,  as.numeric(.)))) %>%

  select(Patient,USI, everything(), -AllFail,-FailedFilterFusions,-alignment_params, -dbsnp, -dgv, -id,
         -senses,-probe, -strands, -genomic_regions,
         -contig, -contig_regions, -contig_size) %>%
  ungroup() %>%
  unique(.) #remove the duplicate entries with different senses/strands, but everyhting else is identical. Note: flanking_pairs are often the ONLY difference between two rows. 
  
head(sense.fmt.filter)
dim(sense.fmt.filter) #5242   33
```

```{r}
# write.csv(sense.fmt.filter, 
#           "TARGET_AML_0531_1031_Relapse_TransAbyss_Fusions_reformatted_FilteredForNBM_10.30.19.csv",
#           row.names = FALSE)
```


# Define the Primary Fusions/Breakpoints per patient 

```{r}
#Helper Functions for Tidy Cleaning
fusionCategory <- function(geneA,geneB){
  fus <- c(geneA,geneB)
  fus <- paste(fus[order(fus)], collapse = "-") #order alphabetically, so that the category includes  reciprocal fusions
}

collapseFusions <- function(breakpoint){
  paste(unique(breakpoint), collapse = "; ")
}

fillNA <- function(x,USI){
  
  USI <- unique(USI)
  x <- x %>% .[!is.na(.)] %>% unique()
  
  if(length(x) < 1){
    x <- ""
  }
  
  # print(x)
  return(x)
}
```


```{r}
getPrimaryFusion <- function(USI,flanking_pairs,sum.break.reads,spanning_reads,in_frame){
  # breakpoint, flanking_pairs,breakpoint_pairs,spanning_reads,frame are all columns from TransAbyss
  # brk.reads <- max(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2)))
  
  if (all(is.na(sum.break.reads))){
    res <- NA
    return(res)
  }
  
  resolveTies <- function(df){
        df <- df %>%
            group_by(primary) %>%
            mutate(rank=rank(flanking_pairs, ties.method = "first")) %>% #if still a tie in flanking pairs counts, choose the first. 
            mutate(rank=ifelse(primary=="alternate", 0, rank)) %>%
            rowwise() %>%
            mutate(reason=sum(reason,rank)) %>%
            ungroup() %>% 
            mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))
        return(df)
    }
  
  #Create boolean vectors for each type of evidence supporting the "primary" fusion. 
  USI <- unique(USI)
  frames <- ifelse(grepl("in", in_frame),4,0) #wieghted so that in frame is always the primary regardless of having fewer counts.
  brk.reads <-  sum.break.reads == max(sum.break.reads)
  spanning.reads <- spanning_reads == max(spanning_reads)
  flanking.reads <- flanking_pairs == max(flanking_pairs)
  
  temp <- list("frames"=frames,"brk.reads"=brk.reads, 
               "spanning.reads"=spanning.reads, "flanking.reads"=flanking.reads)
  
  #Create a temporary data frame. 
  temp <- bind_cols(temp) %>% 
    bind_cols(.,list("flanking_pairs"=flanking_pairs, "sum.break.reads"=sum.break.reads,
                     "spanning_reads"=spanning_reads,"in_frame"=in_frame))
  
  #Will prioritize the in-frame fusions (weighted), then select that one with the highest reads
    temp <- temp %>%
        rowwise() %>%
        mutate(reason=sum(frames,brk.reads,spanning.reads,flanking.reads)) %>%
        ungroup() %>%
        mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))

    # break ties if there are multiple fusions with same #s of supporting reads.
    # for those with highest evidence ("reason"), choose the one with the higher flanking reads
    if (sum(temp$primary == "primary") > 1){
        temp <- resolveTies(temp)
        if (sum(temp$primary == "primary") >1 ){
          # last check for ties, will break dplyr.
          return(temp)
        }
    }

  res <- temp$primary
  
  return(res)
}
```

```{r warning=FALSE}
sense.fmt.primary <- sense.fmt.filter %>% 
  select(everything(), In_frame=frame) %>% #to avoid confusion with an r function called frame()
  rowwise() %>%
  #Fusion category is to encompass both reciprocal fusions under 1 type/name. 
  mutate(Fusion.Category=fusionCategory(X5.gene,X3.gene)) %>% 
  #using sum of the breakpoint pairs. 
  mutate(sum.break.reads=sum(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2))), 
         Group=case_when(
           grepl("BM[0-9]|RO[0-9]|R0[0-9]", USI)~ "NBM",
           grepl("MPN[0-9]", USI) ~ "MPN",
           grepl("sorted", Patient, ignore.case = T) ~ "FlowSorted",
           TRUE ~ "AML")) %>%
  ungroup() %>%

  group_by(Patient,Fusion.Category) %>%
  mutate(Primary.Fusion=getPrimaryFusion(USI=USI,flanking_pairs = flanking_pairs, 
                                         sum.break.reads = sum.break.reads, 
                                         spanning_reads = spanning_reads, in_frame = In_frame)) %>%
  arrange(Patient,Fusion.Category) %>% 
  
  group_by(Patient,Fusion.Category,Primary.Fusion) %>%
  mutate(Alternate.Breakpoints=ifelse(Primary.Fusion=="alternate", 
                                      collapseFusions(breakpoint = breakpoint), NA)) %>%
  ungroup() %>%
  
  group_by(Patient,Fusion.Category) %>%
  mutate(Alternate.Breakpoints=fillNA(Alternate.Breakpoints, USI=USI)) %>% 
  mutate(Alternate.Breakpoints=ifelse(breakpoint==Alternate.Breakpoints,"",Alternate.Breakpoints)) %>%
  ungroup() %>% 
  
  filter(grepl("primary", Primary.Fusion) | grepl("NonePassedFilter|NoneDetected", Fusion)) %>%
  select(Patient,USI,GSC.library,Group,
         Fusion.Category,Fusion,Exons, X5.breakpoint, X3.breakpoint,
         Alternate.Breakpoints, everything(),-Primary.Fusion)

dim(sense.fmt.primary) #4281   40
length(unique(sense.fmt.primary$Patient)) #543

head(sense.fmt.primary)
```

```{r}
table(sense.fmt.primary$Group, useNA = "always")
```


    
#Add Age and Clinical Information     
    
```{r}
sense.fmt.primary <- sense.fmt.primary %>% 
  left_join(.,select(merged,USI, Protocol, Age.in.years, ISCN), 
            by="USI") %>% 
  select(Patient:GSC.library,Protocol:ISCN,everything())


head(sense.fmt.primary)
dim(sense.fmt.primary) #4281   40
```

```{r}
# write.csv(sense.fmt.primary, "TARGET_AML_0531_1031_Relapse_TransAbyss_Fusions_reformatted_FilteredForNBM_PrimaryFusions_10.30.19.csv", row.names = FALSE)
```

```{r}
tab <- table(sense.fmt.primary$Fusion.Category) %>%
  .[order(., decreasing = TRUE)]  %>%
  data.frame()

head(tab,n=15)
```


#Split into Intra/Inter Chromosomal Fusions 

```{r}
inter <- sense.fmt.primary %>% 
  filter(grepl("interchromosomal|NoneDetected|NonePassedFilter", Type)) %>%
  select(-Fusion.Category)

head(inter)
dim(inter) # 1009   39
# write.csv(inter,"TARGET_AML_0531_1031_Relapse_TransAbyss_sense_fusions_reformatted_InterChromosomal_FilteredForNBM_10.30.19.csv",row.names = FALSE)
```

```{r}
intra <- sense.fmt.primary %>% 
  filter(grepl("intrachromosomal|NoneDetected|NonePassedFilter", Type)) %>%
  select(-Fusion.Category)

head(intra)
dim(intra)  #3297   39
# write.csv(intra, "TARGET_AML_0531_1031_Relapse_TransAbyss_sense_fusions_reformatted_IntraChromosomal_FilteredForNBM_10.30.19.csv", row.names = FALSE)
```


#Split into Intra/Inter Chromosomal Fusions and One Row per Patient. 

```{r}
sense.type <- sense.fmt.primary %>%
  group_by(Type,Patient) %>%
  mutate(All.Fusions.Found=collapseRows(Fusion.Category,uniq = TRUE)) %>%
  ungroup()

length(unique(sense.type$Patient))
# head(sense.type)
dim(sense.type)
```

```{r}
FusionGenes <- sense.type %>%
  select(Patient,GSC.library,All.Fusions.Found,Type) %>%
  unique(.) %>%
  spread(Type,All.Fusions.Found) %>%

  mutate(Group=case_when(
           grepl("BM[0-9]|RO[0-9]|R0[0-9]", Patient)~ "NBM",
           grepl("MPN[0-9]", Patient) ~ "MPN",
           grepl("sorted", Patient, ignore.case = T) ~ "FlowSorted",
           TRUE ~ "DS-AML")) %>%
  # mutate_at(vars(interchromosomal:NonePassedFilter), funs(ifelse(!is.na(NoneDetected), "NoneDetected", .))) %>%
  mutate_at(vars(interchromosomal:intrachromosomal), funs(ifelse( (is.na(.) & Group != "NBM"), "NoneDetected",
                                                                  ifelse( (is.na(.) & Group == "NBM"), "NonePassedFilter", .)))) %>%
  select(-NonePassedFilter)

head(FusionGenes,n=20)
dim(FusionGenes) #543   6
```

```{r}
table(FusionGenes$Group, useNA = "always")
```


#Tally the Fusions 

```{r}
createMultiCols <- function(col,split=FALSE,suffix){
  #col is the column with many factors (eg all fusions). character vector lenght 1.
  #suffix is a character string for the suffix on all column names
  #desinged for dplyr bind_cols() function
  
  #example
  # df <- df %>%
  #   bind_cols(.,createMultiCols(.$Fusion, suffix=".sensefusion")) 
  
  if(split){
    groups <- unique(gsub(" ","",unlist(str_split(col, "; "))))
    groups <- groups[groups != ""]
  }else{
    groups <- unique(col)
  }
  list <- lapply(groups, function(x) ifelse(grepl(paste0(x, ";"), col) | grepl(paste0(x, "$"), col) , "Yes",
                                            ifelse(grepl("^$|Unknown",col) | is.na(col), "Unknown","No"))) 
  list.names <- gsub("-", "_",  groups)
  names(list) <- paste0(list.names,suffix)
  
  return(list)
}
```

```{r}
createCols_byType <- function(df,col){
  # print(head(unlist(df[,col])))
  cols <- createMultiCols(col=unlist(df[,col]),split = TRUE,suffix = c(""))  %>%
    bind_cols() %>%
    mutate(Patient=unlist(df[,"Patient"])) %>%
    select(Patient, everything(), -one_of(c("NonePassedFilter","NoneDetected")))
  
  return(cols)
}
```


```{r}
createFreqTable <- function(createMultiCols.df,type){
  df <- t(sapply(createMultiCols.df[,-1], table)) %>%
          as.data.frame() %>% 
          rownames_to_column("Fusion_Category") %>%
          mutate(Percent=(Yes/(Yes+No))*100) %>%
          arrange(desc(Percent))

  if (type == "Exon"){
    df <- df %>%
            separate(Fusion,c("X5.Gene", "X3.Gene"), sep="exon[0-9]{1,2}", remove = FALSE) %>%
            mutate(X5.Exon=gsub("^.+(exon[0-9]{1,2})_.+", "\\1", Fusion),
                 X3.Exon=gsub("^.+_.+(exon[0-9]{1,2})", "\\1", Fusion),
                 X3.Gene=gsub("_", "", X3.Gene)) %>%
            select(Fusion,contains("Gene"),contains("Exon"),everything())

  }else if(type == "Gene"){
    df <- df %>%
            separate(Fusion_Category,c("GeneA", "GeneB"), sep="\\_", remove = FALSE) %>%
            select(Fusion_Category,contains("Gene"),everything())
  }
  
  return(df)
}
```


```{r}
inter.Genes <- createCols_byType(df=FusionGenes, col="interchromosomal") 
dim(inter.Genes) 
```


```{r warning=FALSE}
Inter.G.Tab <- createFreqTable(inter.Genes, type="Gene")
head(Inter.G.Tab)
# write.csv(Inter.G.Tab, "TARGET_AML_0531_1031_TransAbyss_sense_fusions_reformatted_InterChromosomal_FilteredForNBM_GeneLevel_Tables.csv", row.names = FALSE)
```


```{r}
intra.Genes <- createCols_byType(df=FusionGenes, col="intrachromosomal")

dim(intra.Genes) #457 1158
```

```{r warning=FALSE}
Intra.G.Tab <- createFreqTable(intra.Genes, type="Gene")
head(Intra.G.Tab)

# write.csv(Intra.G.Tab, "TARGET_AML_0531_1031_Relapse_TransAbyss_sense_fusions_reformatted_IntraChromosomal_FilteredForNBM_GeneLevel_Tables.csv", row.names = FALSE)
```


#Session Information 

```{r}
sessionInfo()
```

