---
title: 'Fusion Callers Detection Rates'
author: "Jenny Smith"
date: "May 10, 2018"
output: html_document
---



```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.width = 10, fig.height = 10)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2018.02.05_cat_TransAbyss_data/')
```


```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
getwd()
```


#Define Functions to be used

```{r}
calc_sens_spec <- function(confusionMatrix){
    #confusion matrix must be prediced in the rows, row1== pred.neg, row2==pred.pos
    #true values are columns, col1==true.neg, col2==true.pos. 
    spec <- confusionMatrix[1,1]/
      (confusionMatrix[1,1]+confusionMatrix[2,1])
    
    sen <- confusionMatrix[2,2]/
      (confusionMatrix[2,2]+confusionMatrix[1,2])
    
    acc <- (confusionMatrix[2,2] + confusionMatrix[1,1])/sum(confusionMatrix)
    
    
    F1.score <- 2*(confusionMatrix[2,2])/
      (2*confusionMatrix[2,2] + #2*TP
         confusionMatrix[2,1] + #FP
         confusionMatrix[1,2]) #FN
    
    return(list(sensitivity=sen,specificity=spec,accuracy=acc,F1=F1.score))
  }
```


```{r}

Compare_FusionCallers <- function(CDE.col,fusion.name){
  library(dplyr)
  
  calc_sens_spec <- function(confusionMatrix){
    #confusion matrix must be prediced in the rows, row1== pred.neg, row2==pred.pos
    #true values are columns, col1==true.neg, col2==true.pos. 
    spec <- confusionMatrix[1,1]/
      (confusionMatrix[1,1]+confusionMatrix[2,1])
    
    sen <- confusionMatrix[2,2]/
      (confusionMatrix[2,2]+confusionMatrix[1,2])
    
    acc <- (confusionMatrix[2,2] + confusionMatrix[1,1])/sum(confusionMatrix)
    
    
    F1.score <- 2*(confusionMatrix[2,2])/
      (2*confusionMatrix[2,2] + #2*TP
         confusionMatrix[2,1] + #FP
         confusionMatrix[1,2]) #FN
    
    return(list(sensitivity=sen,specificity=spec,accuracy=acc,F1=F1.score))
  }
  
  comp.means <- function(response,predictor,data){
    library(ggpubr)
    form <- paste(response,"~",predictor)
    comp <- compare_means(as.formula(form), data=data)
    return(comp)
  }
  
  
  extract_pval <- function(diff){
    p.vals <-   diff %>%
      rownames_to_column("type") %>%
      filter(grepl("p.format",type)) %>%
      select(-type) %>%
      mutate_all(funs(as.numeric(as.character(.)))) %>%
      mutate_all(funs(round(., digits = 3))) %>%
      gather(Variable,P.Val)
    
    return(p.vals)
  }
  
  summarize.res <- function(df,group,pvals){
    #group is the column to group_by(), character vector 
    #pvals is the output of extract_pval()
    print(group)
    
    summary <- df %>% 
      group_by_at(group) %>%
      
      summarise_if(is.numeric, funs(mean=mean, median=median, max=max, min=min), na.rm = TRUE) %>% 
      gather(var,value,-one_of(group)) %>%
      
      separate(var, c("Variable", "Statistic"), sep="_") %>%
      spread(Statistic, value) %>%
      mutate_if(is.numeric, funs(round(., digits = 1))) %>%
      
      inner_join(., pvals, by="Variable") %>%
      mutate(Fusion_Caller=rep(group,nrow(.))) %>%
      select(Fusion_Caller, Detected=one_of(group), Variable, mean,P.Val, median,max,min) %>%
      arrange(Variable)
    
    return(summary)
    
  }
  
  # ByCyto <- CDE.1031 %>% 
  #   filter_at(vars(CDE.col), all_vars(. != "Unknown")) %>% #remove unknowns for this analysis
  #   mutate(Karyotype=enc2utf8(Karyotype)) %>%
  #   select(USI,Karyotype, FusionName, CDE.col,
  #          Bone.marrow.leukemic.blast.percentage,Peripheral.blasts,wbc1) %>%
  #   inner_join(., select(depth,USI,lib.size), by="USI") 
  # 
  
  #reformat the CDEs
  df <- CDE.1031 %>%
    
    #remove unkowns from the analysis
    filter(Combined.Interchromosomal != "Unknown" | Combined.Intrachromosomal != "Unknown") %>% 
    filter_at(vars(CDE.col), all_vars(. != "Unknown")) %>%
    
    mutate(Called.Inter.Combined.Callers=ifelse(select(., matches(fusion.name)) == "Yes",1,0), 
           Called.Intra.Combined.Callers=ifelse(select(., matches(fusion.name)) == "Yes",1,0)) %>%
    mutate_at(vars(CDE.col),funs(TruePositive=ifelse(. == "Yes", 1,0))) %>%
    mutate(Karyotype=enc2utf8(Karyotype)) %>%
    
    select(USI,Karyotype, matches(fusion.name), CDE.col,
           Bone.marrow.leukemic.blast.percentage,Peripheral.blasts,wbc1) %>%
  
  #Define confusion matrices
  confusionMatrices <- lapply(df[,c("")], 
                              function(x) table(x, df[,CDE.col]))
  
  #Calculate specificity and sensitivity. Format as dataframe. 
  metric <- sapply(confusionMatrices, calc_sens_spec) %>% 
    apply(., 1, function(x) unlist(x)) %>%
    data.frame() %>%
    rownames_to_column("Metric") %>% 
    arrange(desc(sensitivity,specificity))
  
  #reformate the confusion matrices as a dataframe
  CF.reformatted <- lapply(names(confusionMatrices), function(x) data.frame(confusionMatrices[[x]],
                                                                            Fusion_Caller=rep(x,4), 
                                                                            stringsAsFactors = FALSE)) %>%
    bind_rows() %>%
    mutate(Class=ifelse(x==1 & Var2 == "Yes", "True_Positive", 
                        ifelse(x == 0 & Var2 == "Yes", "False_Negative",
                               ifelse(x==1 & Var2 == "No", "False_Positive", "True_Negative")))) %>%
    select(Fusion_Caller,everything(),-x,-Var2) %>%
    spread(Class,Freq) %>%
    select(Fusion_Caller,True_Positive,False_Positive,True_Negative, False_Negative) %>%
    arrange(desc(True_Positive), False_Negative)
  
  df.sub <- df %>% 
    filter_at(vars(CDE.col), all_vars(. == "Yes"))
  
  predictors <- c("Called.TransAbyss","Called.TransAbyss.Filtered", "Called.K.Mungall",
                  "Union.All","Union.TA.STAR", "Union.TA.STAR.Piz")
  responses <-  c("Bone.marrow.leukemic.blast.percentage","Peripheral.blasts","wbc1","lib.size")
  
  #Examine if  the lib size, blast%, wbc is significantly different in the detection groups. 
  diff <- lapply(predictors, function(p) sapply(responses, comp.means, predictor=p, data=df.sub))
  diff <- lapply(diff, function(x) as.data.frame(apply(x, 2, unlist)))
  names(diff) <- predictors
  
  #Summarize the results of the stat tests from "diff" object. 
  summaries <- lapply(1:6, function(i) 
    summarize.res(df=df.sub, group=predictors[i],pvals = extract_pval(diff[[i]]))) %>%
    bind_rows()
  
  res <- list(df,CF.reformatted,metric,summaries)
  names(res) <- c("Analyzed_df","ConfusionMatrices","Detection_Metrics","Covariates_Summaries")
  
  return(res)  
}
```


  PredictedPos <- sense %>%
    mutate(USI=str_split_fixed(Patient, "-", n=3)[,1]) %>%
    filter(grepl(fusion.name, genes)) %>% 
    select(USI) %>%
    unique() %>%
    unlist()
  
  starPredictedPos <- star %>% 
    mutate(USI=str_split_fixed(Patient, "-", n=3)[,1]) %>%
    filter(grepl(fusion.name, X.Fusion)) %>%
    select(USI) %>% 
    unique() %>%
    unlist()
  
  PizPredictedPos <- Pizzly %>% 
    mutate(USI=str_split_fixed(Patient, "-", n=3)[,1]) %>%
    filter(grepl(fusion.name, Fusion)) %>%
    select(USI) %>%
    unique() %>%
    unlist()


#Read in the fusion  data 

```{r}
ITD <- read.csv("TARGET_AML_1031_TransAbyss_ITD.tsv_4.6.18.csv", stringsAsFactors = FALSE, row.names = 1)

ITD$USI <- str_split_fixed(ITD$Patient,"-",n=3)[,1]
ITD$Patient <- gsub("RNA.mRNAseq.level2.fusion.2017July_BCCA_1031_TransAbyss_Illumina_data..", "", ITD$Patient)

# head(ITD)
# dim(ITD) #189,107 by 35
# write.csv(ITD,"TARGET_AML_1031_TransAbyss_ITD.tsv_4.6.18.csv") #quick fix to cell lines IDs
```


```{r}
PTD <- read.csv("TARGET_AML_1031_TransAbyss_PTD.tsv_4.6.18.csv", stringsAsFactors = FALSE, row.names = 1)

PTD$USI <- str_split_fixed(PTD$Patient,"-",n=3)[,1]
PTD$Patient <- gsub("RNA.mRNAseq.level2.fusion.2017July_BCCA_1031_TransAbyss_Illumina_data..", "", PTD$Patient)

# head(PTD)
# dim(PTD) #1,954,176   by   35
# write.csv(PTD,"TARGET_AML_1031_TransAbyss_PTD.tsv_4.6.18.csv")  #quick fix to cell lines IDs
```


#Read in the Clinical Data

```{r}
CDE.1031 <- read.csv("~/reference_mapping-files/TARGET_AML_1031_CDE_cleaned_6.25.18.csv", 
                     stringsAsFactors = FALSE)

CDE.1031 <- CDE.1031 %>%
  filter(ScreenedForFusion=="Yes")

head(CDE.1031[,1:5])
# dim(CDE.1031) # 1037  by 65
```


```{r}
table(CDE.1031$FLT3.ITD.positive.) #168 were positive and had RNAseq.
```


```{r}
FLT3.pos <- CDE.1031 %>%
  filter(grepl("Yes", FLT3.ITD.positive.)) %>%
  select(USI) %>%
  unlist()

length(FLT3.pos) #168
```




#Examine FLT3-ITDs

##Background 

https://www.nature.com/articles/nrc1169
The FLT3 internal tandem duplication (ITD) results from a head-to-tail duplication of 3â€“400 base pairs in *exons 14 or 15*, which encode the juxtamembrane domain of FLT3.

median size is about 40 bp for ITD. 

```{r}
FLT3.ITD <- ITD %>% 
  inner_join(., select(CDE.1031, USI, FLT3.ITD.positive.), by="USI") %>%
  select(Patient,USI,FLT3.ITD.positive.,genes,exons.introns,size) %>%
  unique(.)


FLT3.ITD[1:5,]
# dim(FLT3.ITD) #101500      6
```

There were two true positives only found. 

```{r}
CDE.1031 %>%
  filter(USI %in% subset(FLT3.ITD, grepl("FLT3", FLT3.ITD$genes))$USI) %>%
  select(USI,FLT3.ITD.positive.,FLT3.ITD.allelic.ratio)


FLT3.ITD %>% 
  filter(grepl("FLT3", genes))
  # select(Patient,USI, genes, size, frame)
```

```{r}
FLT3.ITD.rmDups <- FLT3.ITD %>%
  select(USI,FLT3.ITD.positive.,genes) %>%
  unique(.) %>%
  
  dplyr::group_by(USI) %>%
  dplyr::mutate(Called.ITD=ifelse(any(grepl("FLT3", genes)), "Yes", "No")) %>%
  # dplyr::summarise(Called.ITD=ifelse(any(grepl("FLT3", genes)), "Yes", "No"))
  # filter(USI=="PAWVBV")
  select(USI,FLT3.ITD.positive., Called.ITD) %>%
  unique(.)
  

FLT3.ITD.rmDups[1:5,]
```


```{r}
#confusion matrix (CF)
CF.ITD <- table(FLT3.ITD.rmDups$Called.ITD,FLT3.ITD.rmDups$FLT3.ITD.positive.)
```

```{r}
ITD.metrics <- calc_sens_spec(confusionMatrix = CF.ITD)

ITD.metrics #1.19% sensitivity 
```


Both were true positives, but severly underdetected all the rest. 



#Examine FLT3-PTD and MLL-PTD

https://www.ncbi.nlm.nih.gov/pubmed/12529668

Internal tandem duplications (ITD) and D835 point mutations of the receptor tyrosine kinase (RTK) FLT3 are found in a high proportion of cases with acute myeloid leukemia (AML). These genetic aberrations may lead to the constitutive activation of the receptor, thus providing the molecular basis for a persisting growth stimulus. We have screened 69 AML-derived cell lines for FLT3 mutations. 

Four of these cell lines showed ITD of the FLT3 gene, none carried a D835 point mutation. Two cell lines (MUTZ-11 and *MV4-11*) expressed exclusively the mutated allele, the other two cell lines (MOLM-13 and PL-21) displayed a mutated and the wild-type version of the gene. Although mutationally activated FLT3 is supposed to substitute for the stimulatory signal of a growth factor, one of these cell lines (MUTZ-11) was strictly cytokine-dependent. FLT3 transcripts were found in all four cell lines, but the *constitutively phosphorylated receptor protein was clearly detectable only in cell line MV4-11*, possibly explaining why MUTZ-11 cells were growth-factor dependent. 


From TransAbyss Documentation

partial tandem duplication events where the 3' exon boundary of a downstream exon is joined to the 5' exon boundary of an upstream exon within the same gene


```{r}
FLT3.PTD <- PTD %>% 
  inner_join(., select(CDE.1031, USI, FLT3.ITD.positive.), by="USI") %>%
  select(Patient,USI,FLT3.ITD.positive., genes,exons.introns,size,
         frame,matches("pairs|reads")) %>%
  unique(.)

FLT3.PTD[1:5,]
```

```{r}
length(intersect(FLT3.PTD$USI,FLT3.pos)) #168 were found True Positive for ITD
```

  select(USI,FLT3.ITD.positive.,genes) %>%
  unique(.) %>%
  
  dplyr::group_by(USI) %>%
  dplyr::mutate(Called.ITD=ifelse(any(grepl("FLT3", genes)), "Yes", "No")) %>%
  # dplyr::summarise(Called.ITD=ifelse(any(grepl("FLT3", genes)), "Yes", "No"))
  # filter(USI=="PAWVBV")
  select(USI,FLT3.ITD.positive., Called.ITD) %>%
  unique(.)

```{r}
FLT3.PTD.rmDups <- FLT3.PTD %>%
  select(USI,FLT3.ITD.positive., genes) %>%
  unique(.) %>% 
  
  dplyr::group_by(USI) %>%
  dplyr::mutate(Called.FLT3.PTD=ifelse(any(grepl("FLT3", genes)), "Yes", "No")) %>% #note: all in-frame
  select(USI,FLT3.ITD.positive., Called.FLT3.PTD) %>%
  unique(.)


FLT3.PTD.rmDups[1:5,]
```


```{r}
CF.PTD <- table(FLT3.PTD.rmDups$Called.FLT3.PTD,FLT3.PTD.rmDups$FLT3.ITD.positive.) #150 true positives and 610 false positives.
CF.PTD
```

```{r}
calc_sens_spec(CF.PTD)
```

```{r}
#Filtering for only exons 14 and 15 result to increase True positives  possibly

FLT3.PTD.rmDups_Filter <- FLT3.PTD %>%
  select(USI,FLT3.ITD.positive., genes, exons.introns) %>%
  filter(ifelse(grepl("FLT3", genes),  grepl("exon15|exon14", exons.introns), TRUE)) %>%
  unique(.) %>%
  
  # filter(grepl("FLT3", genes))
  
  dplyr::group_by(USI) %>%
  dplyr::mutate(Called.FLT3.PTD=ifelse(any(grepl("FLT3", genes)), "Yes", "No")) %>% #note: all in-frame
  select(USI,FLT3.ITD.positive., Called.FLT3.PTD) %>%
  unique(.)

FLT3.PTD.rmDups_Filter[1:5,]
```


```{r}
CF.Filter.PTD <- table(FLT3.PTD.rmDups_Filter$Called.FLT3.PTD, FLT3.PTD.rmDups_Filter$FLT3.ITD.positive.) #69 were FLT3.ITD true positive out of 315 patients 
CF.Filter.PTD
```

```{r}
calc_sens_spec(CF.Filter.PTD)
```




#Session Information 

```{r}
sessionInfo()
```

