---
title: 'Concatenate TransAbyss Fusions'
author: "Jenny Smith"
date: 'April 1,2018'
output: html_document
---


```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.width = 10, fig.height = 10)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2018.02.05_cat_TransAbyss_data')
```

```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
getwd()
```

```{r}
source("~/scripts/conversion_scripts/Merge_Cat_FixDupIDs_Function.r")
```



#Read in the CDEs 


```{r}
CDE.1031 <- read.csv("~/reference_mapping-files/TARGET_AML_1031_CDE_cleaned_7.06.2018.csv", stringsAsFactors = FALSE)


CDE.1031 <- CDE.1031 %>%
  filter(!is.na(USI))

# dim(CDE.1031) #1149 samples
head(CDE.1031[,1:5])
```


```{r}
merged <- read.csv("~/reference_mapping-files/TARGET_AML_1031_0531_Merged_CDE_5.23.18.csv", stringsAsFactors = FALSE)

merged <- merged %>%
  filter(!is.na(TARGET.USI.1))

# dim(merged) #1160  140
head(merged[,1:5])
```



#Merge the patient level files 

```{r}
# dir.create("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2018.02.05_cat_TransAbyss_data")
# setwd("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2018.02.05_cat_TransAbyss_data")
path="/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/level2/fusion/2017July_BCCA_1031_TransAbyss_Illumina_data/"
```



```{r}
files <- dir(path = path,
             pattern = ".tsv")


fileTypes <- str_split_fixed(files, "_", n=3)
tab <- table(fileTypes[,3])
```

```{r}
forMerge <- lapply(names(tab), function(x) paste(path, subset(files, grepl(paste0("_",x), files)), sep="/"))
names(forMerge) <- names(tab)
summary(forMerge)
```


```{r message=FALSE}
merged.datasets <- lapply(forMerge,catRbind, regex="^.+\\/([PBRS].+)\\_A.+",header=TRUE)
merged.datasets <- lapply(merged.datasets, function(x) select(x, Patient, everything())) #reorder the columns
summary(merged.datasets)
```


```{r}
#Updated the code for catRbind to include a dummy data.frame if there are no called fusions for that patient. 
sapply(merged.datasets,function(x) length(unique(x$Patient)))
lapply(merged.datasets, function(x) head(x[,1:5]))

```

```{r}
sapply(merged.datasets, function(x) dim(x))
```

```{r}
# lapply(names(merged.datasets), function(x) write.csv(merged.datasets[[x]], file = paste0("TARGET_AML_1031_TransAbyss_", x,"_4.6.18.csv")))
```


```{r}
##### Define Functions ################

collapseRows <- function(col, uniq=FALSE){
  #designed for dplyr so that col is the vector of the column name
  if (uniq){col <- unique(col)}
  
  collapsed <- ifelse(all(is.na(col)), NA, paste(col, collapse = "; "))
  return(collapsed)
}



filterFusions <- function(Fusion, breakpoint,df2.Fusion,df2.Breakpoints){

  fus <- paste(Fusion,breakpoint,sep=" ")
  fus2 <- paste(df2.Fusion, df2.Breakpoints, sep=" ")
  
  if (fus %in% fus2){
      res <- 1 #these fusions will be filtered out. 
  }else{
    res <- 0 #these one will not. 
  }
  
  return(res)
}


removeDups <- function(col.rowwise){
  
  uniqued <- str_split(col.rowwise, "; ") %>% 
  unlist() %>%
  gsub(" ","",.) %>%
  unique() %>%
  paste(.,collapse = "; ")
  
  return(uniqued)
  
}


comp.means <- function(response,predictor,data){
  library(ggpubr)
  form <- paste(response,"~",predictor)
  comp <- compare_means(as.formula(form), data=data)
  return(comp)
}


extract_pval <- function(diff){
  p.vals <-   diff %>%
    rownames_to_column("type") %>%
    filter(grepl("p.format",type)) %>%
    select(-type) %>%
    mutate_all(funs(as.numeric(as.character(.)))) %>%
    mutate_all(funs(round(., digits = 3))) %>%
    gather(Variable,P.Val)
  
  return(p.vals)
}


```

```{r}
calc_sens_spec <- function(confusionMatrix){
  #confusion matrix must be prediced in the rows, row1== pred.neg, row2==pred.pos
  #true values are columns, col1==true.neg, col2==true.pos. 
  
  spec <- confusionMatrix[1,1]/
          (confusionMatrix[1,1]+confusionMatrix[2,1])
  
  sen <- confusionMatrix[2,2]/
          (confusionMatrix[2,2]+confusionMatrix[1,2])
  
  acc <- (confusionMatrix[2,2] + confusionMatrix[1,1])/sum(confusionMatrix)


  F1.score <- 2*(confusionMatrix[2,2])/
              (2*confusionMatrix[2,2] + #2*TP
               confusionMatrix[2,1] + #FP
               confusionMatrix[1,2]) #FN
  res <- list(sensitivity=sen,specificity=spec,accuracy=acc,F1=F1.score)
  res <- lapply(res, round, digits=2)
  
  return(res)
}
```


```{r}
summarize.res <- function(df,group,pvals){
  #group is the column to group_by(), character vector 
  #pvals is the output of extract_pval()
  print(group)
  
  summary <- df %>% 
    group_by_at(group) %>%
    
    summarise_if(is.numeric, funs(mean=mean, median=median, max=max, min=min), na.rm = TRUE) %>% 
    gather(var,value,-one_of(group)) %>%
    
    separate(var, c("Variable", "Statistic"), sep="_") %>%
    spread(Statistic, value) %>%
    mutate_if(is.numeric, funs(round(., digits = 1))) %>%
    
    inner_join(., pvals, by="Variable") %>%
    mutate(Fusion_Caller=rep(group,nrow(.))) %>%
    select(Fusion_Caller, Detected=one_of(group), Variable, mean,P.Val, median,max,min) %>%
    arrange(Variable)
  
  return(summary)
  
}
```



#################  Sense Fusions ######################

```{r}
sense <- read.csv("TARGET_AML_1031_TransAbyss_sense_fusion.tsv_4.6.18.csv", 
                  stringsAsFactors = FALSE, row.names = 1) %>%
          mutate(USI=str_split_fixed(Patient, "-", n=3)[,1])

# head(sense)
dim(sense) #47,107 by 35
```

```{r}
sense <- sense %>%
  filter(!grepl("Kas|MV4", Patient)) #remove cell lines

length(unique(sense$Patient)) #1,112
dim(sense) #46,873    36
```


```{r}
#Patients with 1 - 240 maximum # of fusions called. 
#38 fusions is the median per patient
quantile(table(sense$Patient))
```

```{r}
# sapply(sense, function(x) grep("var1", x, value=TRUE))
```


# Columns Description 

*From the documentation. *
 flanking_pairs:	number of read pairs from reads-to-genome alignments with both mates flanking the breakpoint, both pointing towards each other.
 breakpoint_pairs:	number of read pairs from reads-to-genome alignments with one mate spanning the breakpoint and the other mate flanking it, both pointing towards each other.  This is useful for read-support when reads lengths are long compared to fragment size.  Pairs up- and down-stream of the breakpoint are reported in a 2-member tuple.
 spanning_reads:	number of reads spanning junction from reads-to-contigs alignments


1)	There is a column in the files called “size”. This contains the distance between the two breakpoints. For example, for sample PANISJ-09A-01, AKR1C1-AKR1E2 fusion, with breakpoints 10:4950612 - 10:4872867 would have a size of 77,744 bp. This is the size of the event.  
a.	Would you also like the distance between the canonical genes themselves? 
 
2)	For identical fusion partners, with different genomic breakpoints recorded in the same patient, we could collapse those so that we report the one fusion per row, and append a column called “Alternate Breakpoints” and just keep the information of those alt breakpoints in one cell of the spreadsheet. 
a.	The breakpoint reported as the “primary” breakpoint for a patient will be in-frame with the highest reported read counts. 
b.	The Alt. Breakpoints columns  can be later queried to determine if it affects the clinical characteristics.  
c.	 Note there are identical fusions with identical breakpoints reported multiple times still since the “flanking_pairs” column has different numbers of reads. This is due to the fusion contigs created during the analysis which looked for the 4 possible orientations of the fusion (RL,LR,LL,RR).  So some of these had different # of reads aligning to the different contigs for one fusion. 
 
3)	I will add in the gene families in a new column next to the fusion names.  



```{r}
reformat.fusions <- function(df){
  # customCols <- c("Patient", "Fusion","All.Sense.Fusions.Found","All.Fusion.Exons","All.Genomic.Breakpoint.ReadPairs","All.Contig.Spanning.Reads", "All.Reading.Frames","All.BreakPoints", "All.Fusion.Sizes")
  # originalCols <- colnames(df)
  
  
  df <- df %>% 
    arrange(Patient) %>%
    mutate(USI=str_split_fixed(Patient, "-", n=3)[,1]) %>%
    
    mutate(Fusion.Detected.TransAbyss=ifelse(is.na(genes), "NoneDetected","Detected")) %>%
    mutate_at(vars(c("X5.gene","X3.gene")), funs(gsub("-","\\.", .))) %>% #replace dashes with a dot. 
    
    mutate(Fusion=ifelse(! is.na(genes), paste(X5.gene, X3.gene, sep="-"), NA),
           Exons=ifelse(! is.na(genes), paste(paste0(X5.gene,"exon",X5.exon), 
                                              paste0(X3.gene,"exon",X3.exon), sep="-"), NA),
           FusionFrame=ifelse(! is.na(genes), paste(Fusion, frame,"frame", sep="_"), NA),
           Type=ifelse(! is.na(genes), ifelse(grepl("^[0-9]", size), 
                                              "intrachromosomal", "interchromosomal"), NA)) %>%
           # X5.breakpoint=str_split_fixed(breakpoint,"\\|", n=2)[,1], 
           # X3.breakpoint=str_split_fixed(breakpoint,"\\|", n=2)[,2]) %>%
    
    mutate(b1=str_split_fixed(breakpoint,"\\|", n=2)[,1], 
           b2=str_split_fixed(breakpoint,"\\|", n=2)[,2]) %>% 
    mutate(b1=paste(b1,str_split_fixed(genes,",", n=2)[,1], sep="_"), 
           b2=paste(b2,str_split_fixed(genes,",", n=2)[,2], sep="_")) %>% 
    
    #Needed to rearrange breakpoints to be in same orientation at X5.gene, X3. gene
    #Original TransAbyss data has breakpoints sorted by genomic coordinate, not which partenr was 5' or 3'
    rowwise() %>%
    mutate(X5.breakpoint = case_when(
              str_detect(b1, X5.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X5.gene) ~ str_split_fixed(b2, "_",n=2)[,1])) %>% 
    mutate(X3.breakpoint = case_when(
              str_detect(b1, X3.gene) ~ str_split_fixed(b1, "_",n=2)[,1],
              str_detect(b2, X3.gene) ~ str_split_fixed(b2, "_",n=2)[,1])) %>% 
    
    rowwise() %>%
    select(Patient,USI,Fusion,Exons, X5.breakpoint,X3.breakpoint, FusionFrame,everything(),-b1,-b2) %>%
    mutate_at(vars(3:NCOL(.)), funs(as.character(.))) %>%
    mutate_at(vars(3:NCOL(.)), funs(ifelse(grepl("NoneDetected", Fusion.Detected.TransAbyss), 
                                           "NoneDetected", .))) 
   
  
 
  return(df)
}
```


```{r}
sense.fmt <- reformat.fusions(sense)
head(sense.fmt)
```


```{r}
# head(sense.fmt)
length(unique(sense.fmt$Patient)) #1,112 patients
dim(sense.fmt) #46,873   by 47 cols
# write.csv(sense.fmt, "TARGET_AML_1031_TransAbyss_sense_fusions_reformatted_06.05.18.csv", row.names = FALSE)
```


```{r}
# sapply(sense.fmt, function(x) grep("var1", x, value=TRUE))
```


#Filter out Fusions found in NBMs 

```{r}
NBM <- sense.fmt %>%
  filter(grepl("^BM|^RO", USI)) %>%
  mutate_at(vars(3:NCOL(.)),funs(gsub("/", "-", .))) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(3:NCOL(.)),funs(gsub(" ", "", .))) ##to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
```


```{r warning=FALSE}
sense.fmt.filter <- sense.fmt %>%
  mutate_at(vars(3:NCOL(.)),funs(gsub("/", "-", .))) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  mutate_at(vars(3:NCOL(.)),funs(gsub(" ", "", .))) %>% #to address the slash at "TNRC6Cexon3-septin9/TNRC6Cfusionexon1""
  
  rowwise() %>%
  mutate(FailedFilterFusions=filterFusions(Fusion, breakpoint, NBM$Fusion,NBM$breakpoint)) %>%
  ungroup() %>%
    
  group_by(Patient) %>%
  mutate(AllFail=ifelse(sum(FailedFilterFusions) == n(),"Yes","No")) %>%
  mutate(Fusion.Detected.TransAbyss=ifelse((AllFail == "Yes" & Fusion.Detected.TransAbyss != "NoneDetected"),
                                           "NonePassedFilter",Fusion.Detected.TransAbyss)) %>%
  ungroup()  %>%
  
  filter(AllFail == "Yes" | FailedFilterFusions == 0) %>%

  rowwise() %>%
  mutate_all(funs(as.character(.))) %>% #change all to characters so that numeric cols can be filled below with a charcter string.
  mutate_at(vars(3:NCOL(.)), funs(ifelse(Fusion.Detected.TransAbyss == "NonePassedFilter","NonePassedFilter",  .))) %>%
  mutate_at(vars(c("flanking_pairs","spanning_reads")), funs(ifelse(Fusion.Detected.TransAbyss == "NonePassedFilter", NA,  as.numeric(.)))) %>%

  select(Patient,USI, everything(), -AllFail,-FailedFilterFusions,-alignment_params, -dbsnp, -dgv, -id,
         -senses,-probe, -strands, -genomic_regions,
         -contig, -contig_regions, -contig_size) %>%
  unique(.) #remove the duplicate entries with different senses/strands, but everyhting else is identical. Note: flanking_pairs are often the ONLY difference between two rows. 
  
# head(sense.fmt.filter)
dim(sense.fmt.filter) #12,624    32
sense.fmt.filter
```


```{r}
# write.csv(sense.fmt.filter, "TARGET_AML_1031_TransAbyss_Fusions_reformatted_FilteredForNBM_06.05.18.csv", row.names = FALSE)
```

#Define the Primary Fusions/Breakpoints per patient 


```{r}
fusionCategory <- function(geneA,geneB){
  fus <- c(geneA,geneB)
  fus <- paste(fus[order(fus)], collapse = "-") #order alphabetically, so that the category includes  reciprocal fusions
}
```


```{r}
collapseFusions <- function(breakpoint){
  paste(unique(breakpoint), collapse = "; ")
}
```


```{r}
fillNA <- function(x,USI){
  
  USI <- unique(USI)
  x <- x %>% .[!is.na(.)] %>% unique()
  
  if(length(x) < 1){
    x <- ""
  }
  
  # print(x)
  return(x)
}
```


```{r}
getPrimaryFusion <- function(USI,flanking_pairs,sum.break.reads,spanning_reads,in_frame){
  # breakpoint, flanking_pairs,breakpoint_pairs,spanning_reads,frame are all columns from TransAbyss
  # brk.reads <- max(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2)))
  
  if (all(is.na(sum.break.reads))){
    res <- NA
    return(res)
  }
  
  resolveTies <- function(df){
        df <- df %>%
            group_by(primary) %>%
            mutate(rank=rank(flanking_pairs, ties.method = "first")) %>% #if still a tie in flanking pairs counts, choose the first. 
            mutate(rank=ifelse(primary=="alternate", 0, rank)) %>%
            rowwise() %>%
            mutate(reason=sum(reason,rank)) %>%
            ungroup() %>% 
            mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))
        return(df)
    }
  
  #Create boolean vectors for each type of evidence supporting the "primary" fusion. 
  USI <- unique(USI)
  frames <- ifelse(grepl("in", in_frame),4,0) #wieghted so that in frame is always the primary regardless of having fewer counts.
  brk.reads <-  sum.break.reads == max(sum.break.reads)
  spanning.reads <- spanning_reads == max(spanning_reads)
  flanking.reads <- flanking_pairs == max(flanking_pairs)
  
  temp <- list("frames"=frames,"brk.reads"=brk.reads, 
               "spanning.reads"=spanning.reads, "flanking.reads"=flanking.reads)
  
  #Create a temporary data frame. 
  temp <- bind_cols(temp) %>% 
    bind_cols(.,list("flanking_pairs"=flanking_pairs, "sum.break.reads"=sum.break.reads,
                     "spanning_reads"=spanning_reads,"in_frame"=in_frame))
  
  #Will prioritize the in-frame fusions (weighted), then select that one with the highest reads
    temp <- temp %>%
        rowwise() %>%
        mutate(reason=sum(frames,brk.reads,spanning.reads,flanking.reads)) %>%
        ungroup() %>%
        mutate(primary=ifelse(reason == max(reason), "primary", "alternate"))

    # break ties if there are multiple fusions with same #s of supporting reads.
    # for those with highest evidence ("reason"), choose the one with the higher flanking reads
    if (sum(temp$primary == "primary") > 1){
        temp <- resolveTies(temp)
        if (sum(temp$primary == "primary") >1 ){
          # last check for ties, will break dplyr.
          return(temp)
        }
    }

  res <- temp$primary
  
  return(res)
}
```



```{r warning=FALSE}
sense.fmt.primary <- sense.fmt.filter %>% 
  select(everything(), In_frame=frame) %>% #to avoid confusion with an r function called frame()
  rowwise() %>%
  #Fusion category is to encompass both reciprocal fusions under 1 type/name. 
  mutate(Fusion.Category=fusionCategory(X5.gene,X3.gene)) %>% 
  #using sum of the breakpoint pairs. 
  mutate(sum.break.reads=sum(as.numeric(str_split_fixed(breakpoint_pairs,",", n=2))), 
         Group=ifelse(grepl("^BM|RO", USI), "NBM", "AML")) %>%
  ungroup() %>%

  group_by(USI,Fusion.Category) %>%
  mutate(Primary.Fusion=getPrimaryFusion(USI=USI,flanking_pairs = flanking_pairs, sum.break.reads = sum.break.reads, spanning_reads = spanning_reads, in_frame = In_frame)) %>%
  arrange(USI,Fusion.Category) %>% 
  
  group_by(USI,Fusion.Category,Primary.Fusion) %>%
  mutate(Alternate.Breakpoints=ifelse(Primary.Fusion=="alternate", 
                                      collapseFusions(breakpoint = breakpoint), NA)) %>%
  ungroup() %>%
  
  group_by(USI,Fusion.Category) %>%
  mutate(Alternate.Breakpoints=fillNA(Alternate.Breakpoints, USI=USI)) %>% 
  mutate(Alternate.Breakpoints=ifelse(breakpoint==Alternate.Breakpoints,"",Alternate.Breakpoints)) %>%
  ungroup() %>% 
  
  filter(grepl("primary", Primary.Fusion) | grepl("NonePassedFilter|NoneDetected", Fusion)) %>%
  select(Patient,USI,Group,Fusion.Category,Fusion,Exons, X5.breakpoint, X3.breakpoint,Alternate.Breakpoints, everything(),-Primary.Fusion)

dim(sense.fmt.primary) #10,393 by 36
length(unique(sense.fmt.primary$Patient)) #1,112
```


    
#Add Age and Clinical Information     
    
```{r}
sense.fmt.primary <- sense.fmt.primary %>% 
  left_join(., select(merged, TARGET.USI.1, Age.Yrs,ISCN), by=c("USI"="TARGET.USI.1")) %>%
  mutate(Age.Yrs=round(Age.Yrs, digits=3)) %>%
  mutate(Fusion.Category=ifelse(grepl("None", Fusion.Category), "",Fusion.Category)) %>%
  select(Patient,USI,Age.Yrs,ISCN, everything())

# head(sense.fmt.primary)
```

```{r}
# write.csv(sense.fmt.primary, "TARGET_AML_1031_TransAbyss_Fusions_reformatted_FilteredForNBM_PrimaryFusions_06.05.18.csv", row.names = FALSE)
```

```{r}
tab <- table(sense.fmt.primary$Fusion.Category) %>%
  .[order(., decreasing = TRUE)]  %>%
  data.frame()

head(tab)
```


#Split into Intra/Inter Chromosomal Fusions 

```{r}
inter <- sense.fmt.primary %>% 
  filter(grepl("interchromosomal|NoneDetected|NonePassedFilter", Type)) %>%
  select(-Fusion.Category)

head(inter)
# dim(inter) #2,302   37
# write.csv(inter,"TARGET_AML_1031_TransAbyss_sense_fusions_reformatted_InterChromosomal_FilteredForNBM_06.05.18.csv", row.names = FALSE)
```


```{r}
intra <- sense.fmt.primary %>% 
  filter(grepl("intrachromosomal|NoneDetected|NonePassedFilter", Type)) %>%
  select(-Fusion.Category)

head(intra)
# dim(intra) #8,169  37
# write.csv(intra, "TARGET_AML_1031_TransAbyss_sense_fusions_reformatted_IntraChromosomal_FilteredForNBM_06.05.18.csv", row.names = FALSE)
```


#Split into Intra/Inter Chromosomal Fusions and One Row per Patient. 

```{r}
table(sense.fmt.filter$Type)
```

```{r}
table(sense.fmt.primary$Type) #removes dupicate entries, per patient,for the same fusion partners, by 1,563 interchrom and 668 intrachrom. 
```

<!-- interchromosomal intrachromosomal     NoneDetected NonePassedFilter  -->
<!--             2224             8091                5               73  -->

```{r}
sense.type <- sense.fmt.primary %>%
  group_by(Type,Patient) %>%
  mutate(All.Fusions.Found=collapseRows(Fusion.Category,uniq = TRUE)) %>%
  ungroup()

length(unique(sense.type$Patient))
# sense.type
```

```{r}
FusionGenes <- sense.type %>%
  select(USI,All.Fusions.Found,Type) %>%
  unique(.) %>%
  spread(Type,All.Fusions.Found) %>%
  
  mutate(Group=ifelse(grepl("^BM|^RO", USI), "NBM", "AML")) %>% 
  mutate_at(vars(interchromosomal:NonePassedFilter), funs(ifelse(!is.na(NoneDetected), "NoneDetected", .))) %>%
  mutate_at(vars(interchromosomal:intrachromosomal), funs(ifelse( (is.na(.) & Group != "NBM"), "NoneDetected",
                                                                  ifelse( (is.na(.) & Group == "NBM"), "NonePassedFilter", .)))) %>%
  select(-NoneDetected, -NonePassedFilter)

# head(FusionGenes,n=20)
```

```{r}
dim(FusionGenes) 
# write.csv(FusionGenes, "Reformatted_/TARGET_AML_1031_TransAbyss_Fusions_reformatted_FilteredForNBM__ByType_06.05.18.csv")
```


#Tally the Fusions 


```{r}
createMultiCols <- function(col,split=FALSE,suffix){
  #col is the column with many factors (eg all fusions). character vector lenght 1.
  #suffix is a character string for the suffix on all column names
  #desinged for dplyr bind_cols() function
  
  #example
  # df <- df %>%
  #   bind_cols(.,createMultiCols(.$Fusion, suffix=".sensefusion")) 
  
  if(split){
    groups <- unique(gsub(" ","",unlist(str_split(col, "; "))))
    groups <- groups[groups != ""]
  }else{
    groups <- unique(col)
  }
  list <- lapply(groups, function(x) ifelse(grepl(paste0(x, ";"), col) | grepl(paste0(x, "$"), col) , "Yes",
                                            ifelse(grepl("^$|Unknown",col) | is.na(col), "Unknown","No"))) 
  list.names <- gsub("-", "_",  groups)
  names(list) <- paste0(list.names,suffix)
  
  return(list)
}
```



```{r}
createCols_byType <- function(df,col){
  
  # print(head(unlist(df[,col])))
  
  cols <- createMultiCols(col=unlist(df[,col]),split = TRUE,suffix = c(""))  %>%
    bind_cols() %>%
    mutate(USI=unlist(df[,"USI"])) %>%
    select(USI, everything(), -NonePassedFilter, -NoneDetected)
  
  return(cols)
}
```


```{r}
createFreqTable <- function(createMultiCols.df,type){
  
  df <- t(sapply(createMultiCols.df[,-1], table)) %>%
          as.data.frame() %>% 
          rownames_to_column("Fusion_Category") %>%
          mutate(Percent=(Yes/(Yes+No))*100) %>%
          arrange(desc(Percent))

  if (type == "Exon"){
    df <- df %>%
            separate(Fusion,c("X5.Gene", "X3.Gene"), sep="exon[0-9]{1,2}", remove = FALSE) %>%
            mutate(X5.Exon=gsub("^.+(exon[0-9]{1,2})_.+", "\\1", Fusion),
                 X3.Exon=gsub("^.+_.+(exon[0-9]{1,2})", "\\1", Fusion),
                 X3.Gene=gsub("_", "", X3.Gene)) %>%
            select(Fusion,contains("Gene"),contains("Exon"),everything())

  }else if(type == "Gene"){
    df <- df %>%
            separate(Fusion_Category,c("GeneA", "GeneB"), sep="\\_", remove = FALSE) %>%
            select(Fusion_Category,contains("Gene"),everything())
  }
  
  return(df)
}
```


```{r}
inter.Genes <- createCols_byType(df=FusionGenes, col="interchromosomal") 
dim(inter.Genes) #1112 by 1,241
```


```{r warning=FALSE}
Inter.G.Tab <- createFreqTable(inter.Genes, type="Gene")
head(Inter.G.Tab)
# write.csv(Inter.G.Tab, "TARGET_AML_1031_TransAbyss_sense_fusions_reformatted_InterChromosomal_FilteredForNBM_GeneLevel_Tables.csv", row.names = FALSE)
```


```{r}
intra.Genes <- createCols_byType(df=FusionGenes, col="intrachromosomal")

dim(intra.Genes) #1112 by 1,665
```


```{r warning=FALSE}
Intra.G.Tab <- createFreqTable(intra.Genes, type="Gene")
head(Intra.G.Tab)

# write.csv(Intra.G.Tab, "TARGET_AML_1031_TransAbyss_sense_fusions_reformatted_IntraChromosomal_FilteredForNBM_GeneLevel_Tables.csv", row.names = FALSE)
```


#Annotate with Fusions Databases 

## Read in the Fusion Database Information

```{r}
source("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2018.05.15_Intra-Inter_Chromasomal_Fusion_Verification/Query_FusionDBs_for_Aliases.R")
```

```{r}
db_fusionCancer <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/FusionCancer/fusion_table.txt", sep = "\t", header = TRUE)  # Note: had to use the Zap Gremlins feature of BBEdit to remove weird, non-Unicode characters that were present in the original file prior to reading into R
# Creating a Fusion_pair column for easy lookup later
db_fusionCancer$Fusion_pair <- paste(db_fusionCancer$Head_gene_symbol, db_fusionCancer$Tail_gene_symbol, sep = "_")


db_mitelman <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/Mitelman/mitelman_formatted_tidy.txt")
# Creating a Fusion_pair column for easy lookup later
db_mitelman$Fusion_pair <- str_replace(db_mitelman$Gene, pattern = "/", replacement = "_")


db_ticdb <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/TICdb/allseqs_TICdb.txt") # Note: reference IDs can be either from GenBank or Pubmed
# Creating a Fusion_pair column for easy lookup later
db_ticdb$Fusion_pair <- paste(db_ticdb$HGNC_ID_Gene1, db_ticdb$HGNC_ID_Gene2, sep = "_")


# This already has a properly formatted Fusion_pair column, no need to make new columns
db_chimerDB <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/ChimerDB/ChimerDB3.0_ChimerKB.txt")

# COSMIC uses a special translocation notation, requires some manipulation to extract the gene symbols for each fusion gene
db_cosmic <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/COSMIC/CosmicFusionExport.tsv")

# Extracting the gene symbols for each gene involved in the fusion
pattern1 <- "^([[:alnum:]])*(?=\\{)"
cosmic_gene1 <- str_match(as.vector(db_cosmic$Translocation.Name), pattern1)[,1]
pattern2 <- "(?<=_)([[:alnum:]])*(?=\\{)"
cosmic_gene2 <- str_match(as.vector(db_cosmic$Translocation.Name), pattern2)[,1]

# Adding this to the COSMIC database
db_cosmic$Fusion_pair <- paste(cosmic_gene1, cosmic_gene2, sep = "_")

db_tumorFusion <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/TumorFusion/JacksonLab_TumorFusions.txt")
# Creating a Fusion_pair column for easy lookup later
db_tumorFusion$Fusion_pair <- paste(db_tumorFusion$Gene_A, db_tumorFusion$Gene_B, sep = "_")


# File containing Gene Symbols and aliases for each symbol
gene_aliases <- read.delim("/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/0000.00.02_Reference_GeneInfo/Hs_GeneInfo.Symb.Alias.txt", header = FALSE, stringsAsFactors = FALSE)
# Removing unecessary characters
gene_aliases$V2 <- str_replace(gene_aliases$V2, pattern = "-", replacement = "")
```






#Session Information 

```{r}
sessionInfo()
```

