---
title: "Concatenate RNA-seq"
author: "Jenny Smith"
date: "June 6, 2019"
output: html_document
---
  
#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height = 5, fig.width = 5)
knitr::opts_knit$set(root.dir = '/fh/fast/meshinchi_s/workingDir/TARGET/AML_TARGET/RNA/mRNAseq/analysis/2019.09.17_Concatenate_DS_AML_RNAseq/')
```

```{r message=FALSE}
library(dplyr)
library(ggplot2)
library(stringr)
library(reshape2)
getwd()
```

```{r}
source(file.path(SCRIPTS,"conversion_scripts/Merge_Cat_FixDupIDs_Function.r"))
source(file.path(SCRIPTS,"/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/R/rmDupGenes_Function.r"))
```




#Define Functions to be used. 

```{r}
#Function for the TPM conversion. 
# Based on https://groups.google.com/forum/#!topic/rsem-users/W9RQrZIOzA4
#Useage: sapply(cated$RPKM, RPKM_to_TPM)

RPKM_to_TPM <- function(RPKM){
  conversionFactor <- sum(RPKM) / 1E6
  TPM <- RPKM / conversionFactor
  return(TPM)
}
```


```{r}
#https://stackoverflow.com/questions/32040742/dplyrselect-including-all-other-columns-at-end-of-new-data-frame-or-beginni

addCols <- function(df,symbol){
  library(dplyr)
  
  df <- df %>%
    as.data.frame() %>%
    mutate(geneSymbol=symbol) %>%
    select(geneSymbol, everything())
  
  return(df)
}
```


#Download Files from S3

Note this regex works for this data set bc they do not have "TARGET" in thier sample names. 

This would not usually differentiate the samples...

```{bash, eval=FALSE}
ml awscli 
cd /fh/scratch/delete90/meshinchi_s/jlsmith3/kallisto/DSAML
BUCKET="s3://fh-pi-meshinchi-s"
aws s3 cp --only-show-errors --recursive  --exclude "*" --include "P*.h5" --include "P*.json"  $BUCKET/SR/kallisto_out/ .
```



#Identify Files to Be Concatenated

```{r}
filepath <-  file.path(SCRATCH,"jlsmith3/kallisto/DSAML")

allfiles <- dir(path = filepath, pattern = "*.h5",
                full.names = TRUE, recursive = TRUE)

head(allfiles)
```

```{r}
length(allfiles) #228 AML files 
```

```{r}
patient.IDs <- data.frame(filepath=filepath,
                          filename=allfiles) %>% 
  
  mutate(TARGET.Barcode=str_split_fixed(filename, pattern = "[/_]", n=14)[,10]) %>%
  mutate(Lib_Prep="RBS") %>%
  mutate(Final_Colname=gsub("TARGET-[0-9]{2}-","",TARGET.Barcode)) %>% 
  arrange(Final_Colname)


# head(patient.IDs)
dim(patient.IDs) #228   5
# write.csv(patient.IDs, "TARGET_AML_RBD_DS-AML_Sample_IDmap.csv",row.names = FALSE)
```



#Read in a Gene to Transcript ID Map 

https://www.gencodegenes.org/human/stats.html

```{r}
IDmap <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/gencode.v29_RepBase.v24.01_Gene_IDmap.csv")

head(IDmap[,1:5])
dim(IDmap)
```



# TXimport with the HDF5 files

https://wurmlab.github.io/genomicscourse/2016-SIB/practicals/rnaseq/TP2
For this purpose, they introduced the "scaledTPM" values, which are obtained by summing the transcript-level TPMs by gene, and multiplying them with the total library size in millions. 

ScaledTPM values are artificial values, transforming underlying abundance measures to the scale of read counts. This allows to incorporate the information provided by the sequencing depth, and work with RNA-seq differential expression tools that were developed to use read counts.

countsFromAbundance:
character, either "no" (default), "scaledTPM", or "lengthScaledTPM", for whether to generate estimated counts using abundance estimates scaled up to library size (scaledTPM) or additionally scaled using the average transcript length over samples and the library size (lengthScaledTPM). if using scaledTPM or lengthScaledTPM, then the counts are no longer correlated with average transcript length, and so the length offset matrix should not be used.

```{r}
library(rslurm)
```

```{r}
files <-patient.IDs$filename %>% 
  set_names(patient.IDs$Final_Colname)

tx2gene <- dplyr::select(IDmap, transcript_id, gene_id)

sopt <- list(nodes='1', 'cpus-per-task'='16',
             'partition'='largenode', 'mem'='62G',
             'time' = '24:00:00', 'mail-type'='END,FAIL',
             'mail-user'='jlsmith3@fredhutch.org') 

```

```{r}
txi.geneLevel.job <- slurm_call(f=tximport,
                     jobname = "DSAML_gene",
                     params =  list(files = files,
                                    type="kallisto", 
                                    tx2gene = tx2gene, 
                                    txIn = TRUE,
                                    txOut = FALSE,
                                    ignoreAfterBar = TRUE, 
                                    dropInfReps= TRUE,
                                    countsFromAbundance = "scaledTPM"),
                     add_objects = c("files","tx2gene"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 


```

```{r}
get_job_status(txi.geneLevel.job)
```

```{r}
txi.txLevel.job <- slurm_call(f=tximport,
                     jobname = "DSAML_Tx",
                     params =  list(files = files,
                                    type="kallisto", 
                                    tx2gene = tx2gene, 
                                    txIn = TRUE,
                                    txOut = TRUE,
                                    ignoreAfterBar = TRUE, 
                                    dropInfReps= TRUE,
                                    countsFromAbundance = "scaledTPM"),
                     add_objects = c("files","tx2gene"),
                     slurm_options=sopt,
                     submit = TRUE) #Submitted batch job 


# str(txi.txLevel.job) 
```

```{r}
get_job_status(txi.txLevel.job)
```


## read in the results of tximport 

```{r}
txi.geneLevel <- readRDS("_rslurm_DSAML_gene/results_0.RDS") 
txi.geneLevel$countsFromAbundance
```

```{r}
lapply(txi.geneLevel[c(1:3)], function(x) head(x[,1:5]))

sapply(txi.geneLevel[c(1:3)], dim)
```

```{r}
# saveRDS(txi.geneLevel$abundance,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_Abundance_TPM.RDS")
# 
# saveRDS(txi.geneLevel$counts,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_scaledTPM_counts.RDS")
# 
# saveRDS(txi.geneLevel$length,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_GeneLevel_scaledTPM_length.RDS")
```


```{r}
txi.transcriptLevel <- readRDS("_rslurm_DSAML_Tx/results_0.RDS") 
txi.transcriptLevel$countsFromAbundance
```

```{r}
new_rownames <- str_split_fixed(rownames(txi.transcriptLevel$abundance), 
                                pattern = "\\|", n = 2)[,1] 

txi.transcriptLevel[c(1:3)] <- lapply(txi.transcriptLevel[c(1:3)],set_rownames, value=new_rownames)
```

```{r}
lapply(txi.transcriptLevel[c(1:3)], function(x) head(x[,1:5]))
lapply(txi.transcriptLevel[c(1:3)], function(x) tail(x[,1:5]))

sapply(txi.transcriptLevel[c(1:3)], dim)
```

```{r}
# saveRDS(txi.transcriptLevel$abundance,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_Abundance_TPM.RDS")
# 
# saveRDS(txi.transcriptLevel$counts,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_scaledTPM_counts.RDS")
# 
# saveRDS(txi.transcriptLevel$length,
#           "TARGET_AML_RBD_DSAML_Kallisto_Quant_TranscriptLevel_scaledTPM_length.RDS")
```



NOTE:
  
  If markdown cannot be completed with pandoc (memory issues), use the following

> library("rmarkdown")
> render("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/TARGET_AML_1031_Cat_RNAseq.Rmd", "html_document")


#Session Info

```{r}
sessionInfo()
```

